# Task 1b: GraSP Saliency-Based Iterative Pruning

In [7]:
# Setup for Colab and Local environments
import sys
import os

# Check if running on Colab
try:
    import google.colab # type: ignore
    IN_COLAB = True
    print("Running on Google Colab")

    # Mount Google Drive
    from google.colab import drive # type: ignore
    drive.mount('/content/drive')

    # Add project directory to Python path
    # Change this path to match your Google Drive folder structure
    PROJECT_PATH = '/content/drive/MyDrive/AI624-Edge-Devices/pa1.1'
    sys.path.append(PROJECT_PATH)
    print(f"Added {PROJECT_PATH} to Python path")

except ImportError:
    IN_COLAB = False
    print("Running locally")

# Enable PyTorch MPS fallback (macOS only)
from utils import enablePytorchFallback
enablePytorchFallback()

Running on Google Colab
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
Added /content/drive/MyDrive/AI624-Edge-Devices/pa1.1 to Python path


In [8]:
# Core imports
import os
import torch
from utils import getDeviceFromHardware, get_base_path
from task1b import VGG16_GraSP

# Detect device and base path
DEVICE = getDeviceFromHardware()
BASE_PATH = get_base_path(gdrive_relative_path="/content/drive/MyDrive/AI624-Edge-Devices/pa1.1")

print(f"Using device: {DEVICE}")
print(f"Base path: {BASE_PATH}")

Detected Google Colab environment
Mounting Google Drive...
Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
✓ Google Drive mounted successfully
Using specified path: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1
Using device: cuda
Base path: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1


In [9]:
# Create task1b directory structure
TASK1B_DIR = os.path.join(BASE_PATH, 'task1b')
CHECKPOINTS_DIR = os.path.join(TASK1B_DIR, 'checkpoints')
MODELS_DIR = os.path.join(TASK1B_DIR, 'models')
MASKS_DIR = os.path.join(TASK1B_DIR, 'pruning_masks')
RESULTS_DIR = os.path.join(TASK1B_DIR, 'results')

# Create all directories
os.makedirs(CHECKPOINTS_DIR, exist_ok=True)
os.makedirs(MODELS_DIR, exist_ok=True)
os.makedirs(MASKS_DIR, exist_ok=True)
os.makedirs(RESULTS_DIR, exist_ok=True)

print("Task 1b directory structure:")
print(f"  Checkpoints: {CHECKPOINTS_DIR}")
print(f"  Models: {MODELS_DIR}")
print(f"  Masks: {MASKS_DIR}")
print(f"  Results: {RESULTS_DIR}")

Task 1b directory structure:
  Checkpoints: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints
  Models: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/models
  Masks: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/pruning_masks
  Results: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/results


In [10]:
# Target sparsity configuration
TARGET_SPARSITY = 0.80  # 80% pruning (within 70-90% range)

print(f"Target Sparsity: {TARGET_SPARSITY*100}%")
print(f"Stage 1 Sparsity (50% of target): {TARGET_SPARSITY*0.50*100}%")
print(f"Stage 2 Sparsity (75% of target): {TARGET_SPARSITY*0.75*100}%")
print(f"Stage 3 Sparsity (100% of target): {TARGET_SPARSITY*1.00*100}%")

Target Sparsity: 80.0%
Stage 1 Sparsity (50% of target): 40.0%
Stage 2 Sparsity (75% of target): 60.00000000000001%
Stage 3 Sparsity (100% of target): 80.0%


## Part 1: CIFAR-10 Pipeline

In [11]:
# Create VGG16-BN model for CIFAR-10
cifar10_model = VGG16_GraSP(num_classes=10, is_pruned=True)

# Get datasets
train_dataset, test_dataset = cifar10_model.get_train_test_split(BASE_PATH)

# Create data loaders
train_loader, test_loader = cifar10_model.get_data_loaders(train_dataset, test_dataset, batch_size=128)

print(f"CIFAR-10 dataset loaded")
print(f"  Training samples: {len(train_dataset)}")
print(f"  Test samples: {len(test_dataset)}")

Dataset found at /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/datasets/cifar-10-batches-py
CIFAR-10 dataset loaded
  Training samples: 50000
  Test samples: 10000


### Step 2: Run Iterative GraSP Pruning Pipeline

In [12]:
# Run the complete iterative pruning pipeline
# This will:
# 1. Initialize random weights
# 2. Stage 1: Train to 20% → Prune to 40%
# 3. Stage 2: Train to 40% → Prune to 60%
# 4. Stage 3: Train to 60% → Prune to 80%
# 5. Final fine-tuning

cifar10_masks, cifar10_best_acc = cifar10_model.iterative_pruning_pipeline(
    train_loader=train_loader,
    val_loader=test_loader,
    device=DEVICE,
    target_sparsity=TARGET_SPARSITY,
    checkpoint_dir=CHECKPOINTS_DIR,
    masks_dir=MASKS_DIR,
    results_dir=RESULTS_DIR,
    dataset_name='cifar10'
)

print(f"\nCIFAR-10 GraSP Pruning Complete!")
print(f"Best Accuracy: {cifar10_best_acc:.2f}%")


TASK 1B: GRASP ITERATIVE PRUNING PIPELINE (ROBUST MODE)
Dataset: CIFAR10
Target Sparsity: 80.0%
Stage 1 (at 20% acc): 40.0% sparsity (50% of target)
Stage 2 (at 40% acc): 60.00000000000001% sparsity (75% of target)
Stage 3 (at 60% acc): 80.0% sparsity (100% of target)

CHECKING FOR EXISTING CHECKPOINTS (RECOVERY MODE)
  Results JSON:         ✗ Not found
  Final checkpoint:     ✗ Not found
  Stage 3 mask:         ✗ Not found
  Stage 3 checkpoint:   ✗ Not found
  Stage 2 mask:         ✗ Not found
  Stage 2 checkpoint:   ✗ Not found
  Stage 1 mask:         ✗ Not found
  Stage 1 checkpoint:   ✗ Not found

Initializing model with random weights...
✓ Model weights initialized with random values

STAGE 1 TRAINING: Target accuracy 20.0%

Fine-tuning to 20.0% accuracy (max 200 epochs)

Starting training...



Epoch [1/200] Loss: 4.4718 | LR: 0.010000 | Val Top-1: 21.58% | Val Top-5: 80.71%

✓ Target accuracy 20.0% reached! (Val Top-1: 21.58%)
✓ Final checkpoint saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints/cifar10_stage1_trained.pt
Fine-tuning completed. Best accuracy: 21.58%

STAGE 1 PRUNING: Target sparsity 40.0%

GraSP Pruning (Algorithm 1) - Target sparsity: 40.0%
Computing Hessian-gradient product (Algorithm 2)...
  Calibration batches: 4
✓ Hessian-gradient product computed over 4 batches
  Layers processed: 29

Computed threshold: 0.000000
Prunable parameters: 15,244,096
Parameters to keep: 9,146,457 (60.0% of prunable)
Parameters to prune: 6,097,639 (40.0% of prunable)
  features.0: 45.49% sparsity (786/1,728 pruned)
  features.1: 35.94% sparsity (23/64 pruned)
  features.3: 50.21% sparsity (18,508/36,864 pruned)
  features.4: 37.50% sparsity (24/64 pruned)
  features.7: 50.06% sparsity (36,906/73,728 pruned)
  features.8: 28.12% sparsity (36/128 pruned)

Epoch [1/200] Loss: 2.3316 | LR: 0.010000 | Val Top-1: 11.60% | Val Top-5: 53.30%


Epoch [2/200] Loss: 2.3019 | LR: 0.010000 | Val Top-1: 10.00% | Val Top-5: 52.11%


Epoch [3/200] Loss: 2.3007 | LR: 0.010000 | Val Top-1: 11.76% | Val Top-5: 54.54%


Epoch [4/200] Loss: 2.2997 | LR: 0.010000 | Val Top-1: 12.15% | Val Top-5: 56.60%


Epoch [5/200] Loss: 2.2980 | LR: 0.010000 | Val Top-1: 12.60% | Val Top-5: 55.86%


Epoch [6/200] Loss: 2.2964 | LR: 0.010000 | Val Top-1: 13.29% | Val Top-5: 57.65%


Epoch [7/200] Loss: 2.2943 | LR: 0.010000 | Val Top-1: 13.43% | Val Top-5: 56.58%


Epoch [8/200] Loss: 2.2908 | LR: 0.010000 | Val Top-1: 13.71% | Val Top-5: 57.86%


Epoch [9/200] Loss: 2.2869 | LR: 0.010000 | Val Top-1: 14.18% | Val Top-5: 58.57%


Epoch [10/200] Loss: 2.2854 | LR: 0.010000 | Val Top-1: 13.72% | Val Top-5: 57.66%


Epoch [11/200] Loss: 2.2824 | LR: 0.010000 | Val Top-1: 14.14% | Val Top-5: 59.16%


Epoch [12/200] Loss: 2.2748 | LR: 0.010000 | Val Top-1: 14.47% | Val Top-5: 59.45%


Epoch [13/200] Loss: 2.2658 | LR: 0.010000 | Val Top-1: 14.37% | Val Top-5: 60.31%


Epoch [14/200] Loss: 2.2626 | LR: 0.010000 | Val Top-1: 14.56% | Val Top-5: 61.53%


Epoch [15/200] Loss: 2.2368 | LR: 0.010000 | Val Top-1: 17.02% | Val Top-5: 68.48%


Epoch [16/200] Loss: 2.1459 | LR: 0.010000 | Val Top-1: 20.36% | Val Top-5: 82.12%


Epoch [17/200] Loss: 2.0217 | LR: 0.010000 | Val Top-1: 22.36% | Val Top-5: 84.52%


Epoch [18/200] Loss: 1.9645 | LR: 0.010000 | Val Top-1: 22.60% | Val Top-5: 84.24%


Epoch [19/200] Loss: 1.9324 | LR: 0.010000 | Val Top-1: 25.38% | Val Top-5: 85.39%


Epoch [20/200] Loss: 1.9097 | LR: 0.010000 | Val Top-1: 25.69% | Val Top-5: 86.63%


Epoch [21/200] Loss: 1.8871 | LR: 0.010000 | Val Top-1: 27.24% | Val Top-5: 86.87%


Epoch [22/200] Loss: 1.8635 | LR: 0.010000 | Val Top-1: 26.77% | Val Top-5: 86.55%


Epoch [23/200] Loss: 1.8415 | LR: 0.010000 | Val Top-1: 29.51% | Val Top-5: 88.92%


Epoch [24/200] Loss: 1.8102 | LR: 0.010000 | Val Top-1: 29.03% | Val Top-5: 88.49%


Epoch [25/200] Loss: 1.7868 | LR: 0.010000 | Val Top-1: 31.44% | Val Top-5: 90.31%


Epoch [26/200] Loss: 1.7614 | LR: 0.010000 | Val Top-1: 31.78% | Val Top-5: 89.90%


Epoch [27/200] Loss: 1.7392 | LR: 0.010000 | Val Top-1: 33.77% | Val Top-5: 90.50%


Epoch [28/200] Loss: 1.7172 | LR: 0.010000 | Val Top-1: 33.77% | Val Top-5: 90.83%


Epoch [29/200] Loss: 1.7004 | LR: 0.010000 | Val Top-1: 32.75% | Val Top-5: 89.31%


Epoch [30/200] Loss: 1.6840 | LR: 0.010000 | Val Top-1: 37.01% | Val Top-5: 91.73%


Epoch [31/200] Loss: 1.6554 | LR: 0.010000 | Val Top-1: 37.31% | Val Top-5: 91.78%


Epoch [32/200] Loss: 1.6369 | LR: 0.010000 | Val Top-1: 36.78% | Val Top-5: 91.53%


Epoch [33/200] Loss: 1.6177 | LR: 0.010000 | Val Top-1: 37.45% | Val Top-5: 91.82%


Epoch [34/200] Loss: 1.6022 | LR: 0.010000 | Val Top-1: 39.16% | Val Top-5: 92.28%


Epoch [35/200] Loss: 1.5780 | LR: 0.010000 | Val Top-1: 33.48% | Val Top-5: 89.93%


Epoch [36/200] Loss: 1.5623 | LR: 0.010000 | Val Top-1: 40.43% | Val Top-5: 93.35%

✓ Target accuracy 40.0% reached! (Val Top-1: 40.43%)
✓ Final checkpoint saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints/cifar10_stage2_trained.pt
Fine-tuning completed. Best accuracy: 40.43%

STAGE 2 PRUNING: Target sparsity 60.00000000000001%

GraSP Pruning (Algorithm 1) - Target sparsity: 60.00000000000001%
Computing Hessian-gradient product (Algorithm 2)...
  Calibration batches: 4
✓ Hessian-gradient product computed over 4 batches
  Layers processed: 29

Computed threshold: -0.000000
Prunable parameters: 9,146,456
Parameters to keep: 3,658,582 (40.0% of prunable)
Parameters to prune: 5,487,874 (60.0% of prunable)
  features.0: 73.21% sparsity (1,265/1,728 pruned)
  features.1: 73.44% sparsity (47/64 pruned)
  features.3: 74.09% sparsity (27,313/36,864 pruned)
  features.4: 60.94% sparsity (39/64 pruned)
  features.7: 72.65% sparsity (53,564/73,728 pruned)
  features.8: 64

Epoch [1/200] Loss: 2.3824 | LR: 0.010000 | Val Top-1: 11.51% | Val Top-5: 53.90%


Epoch [2/200] Loss: 2.2851 | LR: 0.010000 | Val Top-1: 14.45% | Val Top-5: 61.98%


Epoch [3/200] Loss: 2.1932 | LR: 0.010000 | Val Top-1: 20.16% | Val Top-5: 78.20%


Epoch [4/200] Loss: 2.0018 | LR: 0.010000 | Val Top-1: 21.57% | Val Top-5: 80.41%


Epoch [5/200] Loss: 1.9360 | LR: 0.010000 | Val Top-1: 24.52% | Val Top-5: 85.97%


Epoch [6/200] Loss: 1.8915 | LR: 0.010000 | Val Top-1: 26.38% | Val Top-5: 86.78%


Epoch [7/200] Loss: 1.8528 | LR: 0.010000 | Val Top-1: 26.86% | Val Top-5: 84.27%


Epoch [8/200] Loss: 1.8085 | LR: 0.010000 | Val Top-1: 27.39% | Val Top-5: 87.05%


Epoch [9/200] Loss: 1.7590 | LR: 0.010000 | Val Top-1: 31.24% | Val Top-5: 89.37%


Epoch [10/200] Loss: 1.7129 | LR: 0.010000 | Val Top-1: 35.77% | Val Top-5: 89.69%


Epoch [11/200] Loss: 1.6752 | LR: 0.010000 | Val Top-1: 34.56% | Val Top-5: 92.06%


Epoch [12/200] Loss: 1.6461 | LR: 0.010000 | Val Top-1: 35.78% | Val Top-5: 93.27%


Epoch [13/200] Loss: 1.6142 | LR: 0.010000 | Val Top-1: 36.01% | Val Top-5: 91.56%


Epoch [14/200] Loss: 1.5807 | LR: 0.010000 | Val Top-1: 38.52% | Val Top-5: 93.97%


Epoch [15/200] Loss: 1.5570 | LR: 0.010000 | Val Top-1: 41.49% | Val Top-5: 94.90%


Epoch [16/200] Loss: 1.5288 | LR: 0.010000 | Val Top-1: 43.08% | Val Top-5: 94.42%


Epoch [17/200] Loss: 1.4893 | LR: 0.010000 | Val Top-1: 47.58% | Val Top-5: 95.09%


Epoch [18/200] Loss: 1.4516 | LR: 0.010000 | Val Top-1: 48.79% | Val Top-5: 94.81%


Epoch [19/200] Loss: 1.4102 | LR: 0.010000 | Val Top-1: 51.11% | Val Top-5: 95.83%


Epoch [20/200] Loss: 1.3653 | LR: 0.010000 | Val Top-1: 56.96% | Val Top-5: 96.37%


Epoch [21/200] Loss: 1.3193 | LR: 0.010000 | Val Top-1: 59.23% | Val Top-5: 96.17%


Epoch [22/200] Loss: 1.2710 | LR: 0.010000 | Val Top-1: 61.33% | Val Top-5: 96.38%

✓ Target accuracy 60.0% reached! (Val Top-1: 61.33%)
✓ Final checkpoint saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints/cifar10_stage3_trained.pt
Fine-tuning completed. Best accuracy: 61.33%

STAGE 3 PRUNING: Target sparsity 80.0%

GraSP Pruning (Algorithm 1) - Target sparsity: 80.0%
Computing Hessian-gradient product (Algorithm 2)...
  Calibration batches: 4
✓ Hessian-gradient product computed over 4 batches
  Layers processed: 29

Computed threshold: -0.003190
Prunable parameters: 3,658,581
Parameters to keep: 731,716 (20.0% of prunable)
Parameters to prune: 2,926,865 (80.0% of prunable)
  features.0: 87.91% sparsity (1,519/1,728 pruned)
  features.1: 87.50% sparsity (56/64 pruned)
  features.3: 88.74% sparsity (32,713/36,864 pruned)
  features.4: 75.00% sparsity (48/64 pruned)
  features.7: 87.09% sparsity (64,209/73,728 pruned)
  features.8: 81.25% sparsity (104/128 prune

Epoch [1/200] Loss: 2.4996 | LR: 0.010000 | Val Top-1: 18.27% | Val Top-5: 80.73%


Epoch [2/200] Loss: 2.0121 | LR: 0.010000 | Val Top-1: 26.29% | Val Top-5: 85.65%


Epoch [3/200] Loss: 1.9506 | LR: 0.010000 | Val Top-1: 28.60% | Val Top-5: 86.71%


Epoch [4/200] Loss: 1.9070 | LR: 0.010000 | Val Top-1: 28.26% | Val Top-5: 86.96%


Epoch [5/200] Loss: 1.8762 | LR: 0.010000 | Val Top-1: 31.23% | Val Top-5: 88.01%


Epoch [6/200] Loss: 1.8486 | LR: 0.010000 | Val Top-1: 32.49% | Val Top-5: 88.28%


Epoch [7/200] Loss: 1.8289 | LR: 0.010000 | Val Top-1: 32.33% | Val Top-5: 89.31%


Epoch [8/200] Loss: 1.8093 | LR: 0.010000 | Val Top-1: 34.49% | Val Top-5: 89.81%


Epoch [9/200] Loss: 1.7876 | LR: 0.010000 | Val Top-1: 34.95% | Val Top-5: 89.19%


Epoch [10/200] Loss: 1.7709 | LR: 0.010000 | Val Top-1: 36.85% | Val Top-5: 89.60%


Epoch [11/200] Loss: 1.7562 | LR: 0.010000 | Val Top-1: 37.88% | Val Top-5: 91.57%


Epoch [12/200] Loss: 1.7366 | LR: 0.010000 | Val Top-1: 37.33% | Val Top-5: 91.67%


Epoch [13/200] Loss: 1.7274 | LR: 0.010000 | Val Top-1: 38.62% | Val Top-5: 91.66%


Epoch [14/200] Loss: 1.7077 | LR: 0.010000 | Val Top-1: 39.20% | Val Top-5: 91.77%


Epoch [15/200] Loss: 1.6929 | LR: 0.010000 | Val Top-1: 37.82% | Val Top-5: 90.90%


Epoch [16/200] Loss: 1.6837 | LR: 0.010000 | Val Top-1: 40.15% | Val Top-5: 92.39%


Epoch [17/200] Loss: 1.6700 | LR: 0.010000 | Val Top-1: 40.63% | Val Top-5: 92.75%


Epoch [18/200] Loss: 1.6524 | LR: 0.010000 | Val Top-1: 41.36% | Val Top-5: 92.13%


Epoch [19/200] Loss: 1.6412 | LR: 0.010000 | Val Top-1: 41.53% | Val Top-5: 92.48%


Epoch [20/200] Loss: 1.6404 | LR: 0.010000 | Val Top-1: 41.95% | Val Top-5: 93.52%


Epoch [21/200] Loss: 1.6145 | LR: 0.010000 | Val Top-1: 42.80% | Val Top-5: 93.33%


Epoch [22/200] Loss: 1.6091 | LR: 0.010000 | Val Top-1: 43.15% | Val Top-5: 93.13%


Epoch [23/200] Loss: 1.5939 | LR: 0.010000 | Val Top-1: 43.76% | Val Top-5: 93.81%


Epoch [24/200] Loss: 1.5874 | LR: 0.010000 | Val Top-1: 44.24% | Val Top-5: 93.87%


Epoch [25/200] Loss: 1.5693 | LR: 0.010000 | Val Top-1: 45.28% | Val Top-5: 93.72%


Epoch [26/200] Loss: 1.5587 | LR: 0.010000 | Val Top-1: 46.19% | Val Top-5: 94.69%


Epoch [27/200] Loss: 1.5531 | LR: 0.010000 | Val Top-1: 47.04% | Val Top-5: 94.41%


Epoch [28/200] Loss: 1.5455 | LR: 0.010000 | Val Top-1: 45.46% | Val Top-5: 94.30%


Epoch [29/200] Loss: 1.5242 | LR: 0.010000 | Val Top-1: 48.29% | Val Top-5: 93.41%


Epoch [30/200] Loss: 1.5166 | LR: 0.010000 | Val Top-1: 50.94% | Val Top-5: 94.81%


Epoch [31/200] Loss: 1.5088 | LR: 0.010000 | Val Top-1: 48.83% | Val Top-5: 93.26%


Epoch [32/200] Loss: 1.4983 | LR: 0.010000 | Val Top-1: 49.72% | Val Top-5: 94.57%


Epoch [33/200] Loss: 1.4954 | LR: 0.010000 | Val Top-1: 48.33% | Val Top-5: 94.93%


Epoch [34/200] Loss: 1.4799 | LR: 0.010000 | Val Top-1: 48.98% | Val Top-5: 95.48%


Epoch [35/200] Loss: 1.4765 | LR: 0.010000 | Val Top-1: 51.74% | Val Top-5: 94.89%


Epoch [36/200] Loss: 1.4693 | LR: 0.010000 | Val Top-1: 47.84% | Val Top-5: 93.39%


Epoch [37/200] Loss: 1.4600 | LR: 0.010000 | Val Top-1: 51.54% | Val Top-5: 95.17%


Epoch [38/200] Loss: 1.4514 | LR: 0.010000 | Val Top-1: 50.15% | Val Top-5: 94.74%


Epoch [39/200] Loss: 1.4506 | LR: 0.010000 | Val Top-1: 51.89% | Val Top-5: 96.07%


Epoch [40/200] Loss: 1.4388 | LR: 0.010000 | Val Top-1: 51.51% | Val Top-5: 95.97%


Epoch [41/200] Loss: 1.4312 | LR: 0.010000 | Val Top-1: 51.19% | Val Top-5: 95.40%


Epoch [42/200] Loss: 1.4255 | LR: 0.010000 | Val Top-1: 52.98% | Val Top-5: 95.92%


Epoch [43/200] Loss: 1.4251 | LR: 0.010000 | Val Top-1: 54.10% | Val Top-5: 96.04%


Epoch [44/200] Loss: 1.4054 | LR: 0.010000 | Val Top-1: 52.25% | Val Top-5: 95.49%


Epoch [45/200] Loss: 1.4067 | LR: 0.010000 | Val Top-1: 54.21% | Val Top-5: 96.19%


Epoch [46/200] Loss: 1.3971 | LR: 0.010000 | Val Top-1: 55.46% | Val Top-5: 95.62%


Epoch [47/200] Loss: 1.3865 | LR: 0.010000 | Val Top-1: 54.07% | Val Top-5: 95.70%


Epoch [48/200] Loss: 1.3882 | LR: 0.010000 | Val Top-1: 55.36% | Val Top-5: 96.05%


Epoch [49/200] Loss: 1.3725 | LR: 0.010000 | Val Top-1: 56.04% | Val Top-5: 96.40%


Epoch [50/200] Loss: 1.3663 | LR: 0.010000 | Val Top-1: 56.06% | Val Top-5: 95.05%


Epoch [51/200] Loss: 1.3636 | LR: 0.010000 | Val Top-1: 53.63% | Val Top-5: 94.61%


Epoch [52/200] Loss: 1.3531 | LR: 0.010000 | Val Top-1: 57.00% | Val Top-5: 96.55%


Epoch [53/200] Loss: 1.3499 | LR: 0.010000 | Val Top-1: 58.53% | Val Top-5: 96.44%


Epoch [54/200] Loss: 1.3429 | LR: 0.010000 | Val Top-1: 56.53% | Val Top-5: 94.74%


Epoch [55/200] Loss: 1.3344 | LR: 0.010000 | Val Top-1: 58.58% | Val Top-5: 96.21%


Epoch [56/200] Loss: 1.3313 | LR: 0.010000 | Val Top-1: 58.75% | Val Top-5: 96.16%


Epoch [57/200] Loss: 1.3251 | LR: 0.010000 | Val Top-1: 60.41% | Val Top-5: 96.70%


Epoch [58/200] Loss: 1.3190 | LR: 0.010000 | Val Top-1: 57.31% | Val Top-5: 96.69%


Epoch [59/200] Loss: 1.3102 | LR: 0.010000 | Val Top-1: 59.16% | Val Top-5: 96.07%


Epoch [60/200] Loss: 1.2940 | LR: 0.002000 | Val Top-1: 60.28% | Val Top-5: 95.68%


Epoch [61/200] Loss: 1.2484 | LR: 0.002000 | Val Top-1: 63.51% | Val Top-5: 97.28%


Epoch [62/200] Loss: 1.2174 | LR: 0.002000 | Val Top-1: 64.21% | Val Top-5: 97.30%


Epoch [63/200] Loss: 1.2120 | LR: 0.002000 | Val Top-1: 64.18% | Val Top-5: 97.39%


Epoch [64/200] Loss: 1.1976 | LR: 0.002000 | Val Top-1: 64.73% | Val Top-5: 97.65%


Epoch [65/200] Loss: 1.1947 | LR: 0.002000 | Val Top-1: 65.27% | Val Top-5: 97.56%


Epoch [66/200] Loss: 1.1826 | LR: 0.002000 | Val Top-1: 65.23% | Val Top-5: 97.67%


Epoch [67/200] Loss: 1.1837 | LR: 0.002000 | Val Top-1: 65.63% | Val Top-5: 97.58%


Epoch [68/200] Loss: 1.1808 | LR: 0.002000 | Val Top-1: 65.30% | Val Top-5: 97.53%


Epoch [69/200] Loss: 1.1720 | LR: 0.002000 | Val Top-1: 65.88% | Val Top-5: 97.54%


Epoch [70/200] Loss: 1.1677 | LR: 0.002000 | Val Top-1: 66.22% | Val Top-5: 97.41%


Epoch [71/200] Loss: 1.1657 | LR: 0.002000 | Val Top-1: 66.55% | Val Top-5: 97.57%


Epoch [72/200] Loss: 1.1582 | LR: 0.002000 | Val Top-1: 65.71% | Val Top-5: 97.53%


Epoch [73/200] Loss: 1.1551 | LR: 0.002000 | Val Top-1: 66.73% | Val Top-5: 97.44%


Epoch [74/200] Loss: 1.1536 | LR: 0.002000 | Val Top-1: 66.31% | Val Top-5: 97.79%


Epoch [75/200] Loss: 1.1448 | LR: 0.002000 | Val Top-1: 67.06% | Val Top-5: 97.59%


Epoch [76/200] Loss: 1.1470 | LR: 0.002000 | Val Top-1: 66.60% | Val Top-5: 97.87%


Epoch [77/200] Loss: 1.1420 | LR: 0.002000 | Val Top-1: 66.65% | Val Top-5: 97.56%


Epoch [78/200] Loss: 1.1399 | LR: 0.002000 | Val Top-1: 67.03% | Val Top-5: 97.65%


Epoch [79/200] Loss: 1.1339 | LR: 0.002000 | Val Top-1: 66.70% | Val Top-5: 97.65%


Epoch [80/200] Loss: 1.1269 | LR: 0.002000 | Val Top-1: 66.75% | Val Top-5: 97.68%


Epoch [81/200] Loss: 1.1303 | LR: 0.002000 | Val Top-1: 66.27% | Val Top-5: 97.55%


Epoch [82/200] Loss: 1.1256 | LR: 0.002000 | Val Top-1: 67.56% | Val Top-5: 97.80%


Epoch [83/200] Loss: 1.1251 | LR: 0.002000 | Val Top-1: 67.43% | Val Top-5: 97.28%


Epoch [84/200] Loss: 1.1162 | LR: 0.002000 | Val Top-1: 67.61% | Val Top-5: 97.25%


Epoch [85/200] Loss: 1.1166 | LR: 0.002000 | Val Top-1: 67.25% | Val Top-5: 97.64%


Epoch [86/200] Loss: 1.1060 | LR: 0.002000 | Val Top-1: 67.79% | Val Top-5: 97.66%


Epoch [87/200] Loss: 1.1095 | LR: 0.002000 | Val Top-1: 67.22% | Val Top-5: 97.53%


Epoch [88/200] Loss: 1.1109 | LR: 0.002000 | Val Top-1: 67.53% | Val Top-5: 97.55%


Epoch [89/200] Loss: 1.1058 | LR: 0.002000 | Val Top-1: 67.60% | Val Top-5: 97.85%


Epoch [90/200] Loss: 1.1030 | LR: 0.002000 | Val Top-1: 67.59% | Val Top-5: 97.70%


Epoch [91/200] Loss: 1.1011 | LR: 0.002000 | Val Top-1: 67.87% | Val Top-5: 97.63%


Epoch [92/200] Loss: 1.0974 | LR: 0.002000 | Val Top-1: 67.51% | Val Top-5: 97.59%


Epoch [93/200] Loss: 1.0998 | LR: 0.002000 | Val Top-1: 68.28% | Val Top-5: 97.29%


Epoch [94/200] Loss: 1.0946 | LR: 0.002000 | Val Top-1: 68.21% | Val Top-5: 97.63%


Epoch [95/200] Loss: 1.0914 | LR: 0.002000 | Val Top-1: 68.01% | Val Top-5: 97.51%


Epoch [96/200] Loss: 1.0925 | LR: 0.002000 | Val Top-1: 68.02% | Val Top-5: 97.63%


Epoch [97/200] Loss: 1.0870 | LR: 0.002000 | Val Top-1: 68.25% | Val Top-5: 97.41%


Epoch [98/200] Loss: 1.0810 | LR: 0.002000 | Val Top-1: 67.86% | Val Top-5: 97.51%


Epoch [99/200] Loss: 1.0834 | LR: 0.002000 | Val Top-1: 68.52% | Val Top-5: 97.51%


Epoch [100/200] Loss: 1.0764 | LR: 0.002000 | Val Top-1: 67.88% | Val Top-5: 97.62%


Epoch [101/200] Loss: 1.0750 | LR: 0.002000 | Val Top-1: 69.08% | Val Top-5: 97.40%


Epoch [102/200] Loss: 1.0807 | LR: 0.002000 | Val Top-1: 68.63% | Val Top-5: 97.54%


Epoch [103/200] Loss: 1.0699 | LR: 0.002000 | Val Top-1: 68.25% | Val Top-5: 97.59%


Epoch [104/200] Loss: 1.0642 | LR: 0.002000 | Val Top-1: 68.84% | Val Top-5: 97.39%


Epoch [105/200] Loss: 1.0668 | LR: 0.002000 | Val Top-1: 68.78% | Val Top-5: 97.75%


Epoch [106/200] Loss: 1.0593 | LR: 0.002000 | Val Top-1: 68.14% | Val Top-5: 97.59%


Epoch [107/200] Loss: 1.0657 | LR: 0.002000 | Val Top-1: 68.17% | Val Top-5: 97.74%


Epoch [108/200] Loss: 1.0583 | LR: 0.002000 | Val Top-1: 68.81% | Val Top-5: 97.18%


Epoch [109/200] Loss: 1.0560 | LR: 0.002000 | Val Top-1: 69.42% | Val Top-5: 97.73%


Epoch [110/200] Loss: 1.0583 | LR: 0.002000 | Val Top-1: 68.58% | Val Top-5: 97.59%


Epoch [111/200] Loss: 1.0605 | LR: 0.002000 | Val Top-1: 68.32% | Val Top-5: 97.51%


Epoch [112/200] Loss: 1.0568 | LR: 0.002000 | Val Top-1: 68.28% | Val Top-5: 97.37%


Epoch [113/200] Loss: 1.0555 | LR: 0.002000 | Val Top-1: 68.64% | Val Top-5: 97.46%


Epoch [114/200] Loss: 1.0578 | LR: 0.002000 | Val Top-1: 68.31% | Val Top-5: 97.58%


Epoch [115/200] Loss: 1.0480 | LR: 0.002000 | Val Top-1: 68.71% | Val Top-5: 97.43%


Epoch [116/200] Loss: 1.0461 | LR: 0.002000 | Val Top-1: 68.12% | Val Top-5: 97.48%


Epoch [117/200] Loss: 1.0477 | LR: 0.002000 | Val Top-1: 69.98% | Val Top-5: 97.55%


Epoch [118/200] Loss: 1.0457 | LR: 0.002000 | Val Top-1: 68.70% | Val Top-5: 97.64%


Epoch [119/200] Loss: 1.0374 | LR: 0.002000 | Val Top-1: 69.35% | Val Top-5: 97.65%


Epoch [120/200] Loss: 1.0405 | LR: 0.000400 | Val Top-1: 69.51% | Val Top-5: 97.47%


Epoch [121/200] Loss: 1.0188 | LR: 0.000400 | Val Top-1: 69.94% | Val Top-5: 97.74%


Epoch [122/200] Loss: 1.0117 | LR: 0.000400 | Val Top-1: 70.16% | Val Top-5: 97.66%


Epoch [123/200] Loss: 1.0026 | LR: 0.000400 | Val Top-1: 70.30% | Val Top-5: 97.70%


Epoch [124/200] Loss: 0.9987 | LR: 0.000400 | Val Top-1: 70.14% | Val Top-5: 97.78%


Epoch [125/200] Loss: 0.9995 | LR: 0.000400 | Val Top-1: 70.40% | Val Top-5: 97.79%


Epoch [126/200] Loss: 1.0028 | LR: 0.000400 | Val Top-1: 70.20% | Val Top-5: 97.66%


Epoch [127/200] Loss: 0.9998 | LR: 0.000400 | Val Top-1: 70.19% | Val Top-5: 97.77%


Epoch [128/200] Loss: 0.9899 | LR: 0.000400 | Val Top-1: 70.56% | Val Top-5: 97.81%


Epoch [129/200] Loss: 0.9938 | LR: 0.000400 | Val Top-1: 70.59% | Val Top-5: 97.74%


Epoch [130/200] Loss: 0.9899 | LR: 0.000400 | Val Top-1: 70.38% | Val Top-5: 97.72%


Epoch [131/200] Loss: 0.9964 | LR: 0.000400 | Val Top-1: 70.42% | Val Top-5: 97.81%


Epoch [132/200] Loss: 0.9938 | LR: 0.000400 | Val Top-1: 70.53% | Val Top-5: 97.78%


Epoch [133/200] Loss: 0.9916 | LR: 0.000400 | Val Top-1: 70.57% | Val Top-5: 97.71%


Epoch [134/200] Loss: 0.9839 | LR: 0.000400 | Val Top-1: 70.40% | Val Top-5: 97.78%


Epoch [135/200] Loss: 0.9883 | LR: 0.000400 | Val Top-1: 70.41% | Val Top-5: 97.73%


Epoch [136/200] Loss: 0.9878 | LR: 0.000400 | Val Top-1: 70.72% | Val Top-5: 97.80%


Epoch [137/200] Loss: 0.9830 | LR: 0.000400 | Val Top-1: 70.61% | Val Top-5: 97.76%


Epoch [138/200] Loss: 0.9814 | LR: 0.000400 | Val Top-1: 70.53% | Val Top-5: 97.89%


Epoch [139/200] Loss: 0.9855 | LR: 0.000400 | Val Top-1: 70.79% | Val Top-5: 97.83%


Epoch [140/200] Loss: 0.9775 | LR: 0.000400 | Val Top-1: 70.84% | Val Top-5: 97.62%


Epoch [141/200] Loss: 0.9831 | LR: 0.000400 | Val Top-1: 70.54% | Val Top-5: 97.71%


Epoch [142/200] Loss: 0.9839 | LR: 0.000400 | Val Top-1: 70.82% | Val Top-5: 97.68%


Epoch [143/200] Loss: 0.9747 | LR: 0.000400 | Val Top-1: 70.71% | Val Top-5: 97.79%


Epoch [144/200] Loss: 0.9795 | LR: 0.000400 | Val Top-1: 70.67% | Val Top-5: 97.77%


Epoch [145/200] Loss: 0.9776 | LR: 0.000400 | Val Top-1: 70.51% | Val Top-5: 97.73%


Epoch [146/200] Loss: 0.9763 | LR: 0.000400 | Val Top-1: 70.59% | Val Top-5: 97.76%


Epoch [147/200] Loss: 0.9724 | LR: 0.000400 | Val Top-1: 70.53% | Val Top-5: 97.77%


Epoch [148/200] Loss: 0.9699 | LR: 0.000400 | Val Top-1: 70.68% | Val Top-5: 97.72%


Epoch [149/200] Loss: 0.9688 | LR: 0.000400 | Val Top-1: 70.66% | Val Top-5: 97.60%


Epoch [150/200] Loss: 0.9757 | LR: 0.000400 | Val Top-1: 70.66% | Val Top-5: 97.74%


Epoch [151/200] Loss: 0.9723 | LR: 0.000400 | Val Top-1: 70.83% | Val Top-5: 97.67%


Epoch [152/200] Loss: 0.9727 | LR: 0.000400 | Val Top-1: 70.38% | Val Top-5: 97.72%


Epoch [153/200] Loss: 0.9731 | LR: 0.000400 | Val Top-1: 70.79% | Val Top-5: 97.78%


Epoch [154/200] Loss: 0.9717 | LR: 0.000400 | Val Top-1: 70.83% | Val Top-5: 97.67%


Epoch [155/200] Loss: 0.9673 | LR: 0.000400 | Val Top-1: 70.93% | Val Top-5: 97.63%


Epoch [156/200] Loss: 0.9697 | LR: 0.000400 | Val Top-1: 70.87% | Val Top-5: 97.72%


Epoch [157/200] Loss: 0.9721 | LR: 0.000400 | Val Top-1: 70.97% | Val Top-5: 97.67%


Epoch [158/200] Loss: 0.9658 | LR: 0.000400 | Val Top-1: 71.10% | Val Top-5: 97.73%


Epoch [159/200] Loss: 0.9689 | LR: 0.000400 | Val Top-1: 71.06% | Val Top-5: 97.60%


Epoch [160/200] Loss: 0.9644 | LR: 0.000080 | Val Top-1: 71.44% | Val Top-5: 97.69%


Epoch [161/200] Loss: 0.9588 | LR: 0.000080 | Val Top-1: 71.14% | Val Top-5: 97.72%


Epoch [162/200] Loss: 0.9597 | LR: 0.000080 | Val Top-1: 71.12% | Val Top-5: 97.68%


Epoch [163/200] Loss: 0.9603 | LR: 0.000080 | Val Top-1: 70.91% | Val Top-5: 97.70%


Epoch [164/200] Loss: 0.9572 | LR: 0.000080 | Val Top-1: 71.15% | Val Top-5: 97.67%


Epoch [165/200] Loss: 0.9578 | LR: 0.000080 | Val Top-1: 71.27% | Val Top-5: 97.64%


Epoch [166/200] Loss: 0.9543 | LR: 0.000080 | Val Top-1: 71.25% | Val Top-5: 97.63%


Epoch [167/200] Loss: 0.9586 | LR: 0.000080 | Val Top-1: 71.25% | Val Top-5: 97.61%


Epoch [168/200] Loss: 0.9588 | LR: 0.000080 | Val Top-1: 71.04% | Val Top-5: 97.64%


Epoch [169/200] Loss: 0.9554 | LR: 0.000080 | Val Top-1: 70.95% | Val Top-5: 97.64%


Epoch [170/200] Loss: 0.9588 | LR: 0.000080 | Val Top-1: 71.23% | Val Top-5: 97.63%


Epoch [171/200] Loss: 0.9491 | LR: 0.000080 | Val Top-1: 70.97% | Val Top-5: 97.69%


Epoch [172/200] Loss: 0.9557 | LR: 0.000080 | Val Top-1: 71.35% | Val Top-5: 97.67%


Epoch [173/200] Loss: 0.9573 | LR: 0.000080 | Val Top-1: 71.23% | Val Top-5: 97.69%


Epoch [174/200] Loss: 0.9515 | LR: 0.000080 | Val Top-1: 71.06% | Val Top-5: 97.62%


Epoch [175/200] Loss: 0.9543 | LR: 0.000080 | Val Top-1: 71.33% | Val Top-5: 97.65%

Early stopping: No improvement for 15 epochs
Best accuracy: 71.44%
Fine-tuning completed. Best accuracy: 71.44%

EVALUATING FINAL MODEL




✓ Results saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/results/cifar10_grasp_results.json

TASK 1B PIPELINE COMPLETE
Final Top-1 Accuracy: 71.33%
Final Top-5 Accuracy: 97.65%
Final Sparsity: 95.20%
Model Size: 58.25 MB


CIFAR-10 GraSP Pruning Complete!
Best Accuracy: 71.44%


### Step 3: COO Sparse Conversion and Profiling

In [13]:
# Convert to COO sparse format (inherited from VGG16_Pruning)
cifar10_sparse_tensors, cifar10_memory_stats = cifar10_model.convert_to_coo_sparse(cifar10_masks)

# Verify consistency
cifar10_model.verify_mask_coo_consistency(cifar10_masks, cifar10_sparse_tensors)


CONVERTING TO COO SPARSE FORMAT

features.0:
  Total params: 1,728
  Non-zeros: 209
  Sparsity: 87.91%
  Dense size: 0.01 MB
  Sparse size: 0.01 MB
  Compression: 0.92x

features.3:
  Total params: 36,864
  Non-zeros: 4,151
  Sparsity: 88.74%
  Dense size: 0.14 MB
  Sparse size: 0.14 MB
  Compression: 0.99x

features.7:
  Total params: 73,728
  Non-zeros: 9,519
  Sparsity: 87.09%
  Dense size: 0.28 MB
  Sparse size: 0.33 MB
  Compression: 0.86x

features.10:
  Total params: 147,456
  Non-zeros: 17,797
  Sparsity: 87.93%
  Dense size: 0.56 MB
  Sparse size: 0.61 MB
  Compression: 0.92x

features.14:
  Total params: 294,912
  Non-zeros: 27,222
  Sparsity: 90.77%
  Dense size: 1.12 MB
  Sparse size: 0.93 MB
  Compression: 1.20x

features.17:
  Total params: 589,824
  Non-zeros: 39,819
  Sparsity: 93.25%
  Dense size: 2.25 MB
  Sparse size: 1.37 MB
  Compression: 1.65x

features.20:
  Total params: 589,824
  Non-zeros: 54,813
  Sparsity: 90.71%
  Dense size: 2.25 MB
  Sparse size: 1.88 MB

{'features.0': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 1728},
 'features.3': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 36864},
 'features.7': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 73728},
 'features.10': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 147456},
 'features.14': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 294912},
 'features.17': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 589824},
 'features.20': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 589824},
 'features.24': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 1179648},
 'features.27': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 2359296},
 'features.30': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 2359296},
 'features.34': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 2359296},
 'features.37': {'consis

In [14]:
# Profile sparse model
cifar10_model.profile_sparse_model(
    sparse_tensors=cifar10_sparse_tensors,
    test_loader=test_loader,
    train_loader=train_loader,
    device=DEVICE
)


PROFILING SPARSE MODEL


REPLACING DENSE WEIGHTS WITH SPARSE TENSORS

✓ features.0: Replaced with sparse tensor (COO format)
✓ features.3: Replaced with sparse tensor (COO format)
✓ features.7: Replaced with sparse tensor (COO format)
✓ features.10: Replaced with sparse tensor (COO format)
✓ features.14: Replaced with sparse tensor (COO format)
✓ features.17: Replaced with sparse tensor (COO format)
✓ features.20: Replaced with sparse tensor (COO format)
✓ features.24: Replaced with sparse tensor (COO format)
✓ features.27: Replaced with sparse tensor (COO format)
✓ features.30: Replaced with sparse tensor (COO format)
✓ features.34: Replaced with sparse tensor (COO format)
✓ features.37: Replaced with sparse tensor (COO format)
✓ features.40: Replaced with sparse tensor (COO format)
✓ classifier.0: Replaced with sparse tensor (COO format)
✓ classifier.3: Replaced with sparse tensor (COO format)
✓ classifier.6: Replaced with sparse tensor (COO format)

✓ All weights replaced with spar

### Step 4: Save Final Model

In [15]:
# Save final pruned model
cifar10_final_path = os.path.join(MODELS_DIR, 'task1b_cifar10_final.pt')
cifar10_model.save(cifar10_final_path)

print(f"CIFAR-10 final model saved to: {cifar10_final_path}")

Saving model to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/models/task1b_cifar10_final.pt
Model saved successfully
CIFAR-10 final model saved to: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/models/task1b_cifar10_final.pt


## Part 2: CIFAR-100 Pipeline

In [16]:
# Create VGG16-BN model for CIFAR-100
cifar100_model = VGG16_GraSP(num_classes=100, is_pruned=True)

# Get datasets
train_dataset_100, test_dataset_100 = cifar100_model.get_train_test_split(BASE_PATH)

# Create data loaders
train_loader_100, test_loader_100 = cifar100_model.get_data_loaders(train_dataset_100, test_dataset_100, batch_size=128)

print(f"CIFAR-100 dataset loaded")
print(f"  Training samples: {len(train_dataset_100)}")
print(f"  Test samples: {len(test_dataset_100)}")

Dataset found at /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/datasets/cifar-100-python
CIFAR-100 dataset loaded
  Training samples: 50000
  Test samples: 10000


### Step 2: Run Iterative GraSP Pruning Pipeline

In [17]:
# Run the complete iterative pruning pipeline for CIFAR-100
cifar100_masks, cifar100_best_acc = cifar100_model.iterative_pruning_pipeline(
    train_loader=train_loader_100,
    val_loader=test_loader_100,
    device=DEVICE,
    target_sparsity=TARGET_SPARSITY,
    checkpoint_dir=CHECKPOINTS_DIR,
    masks_dir=MASKS_DIR,
    results_dir=RESULTS_DIR,
    dataset_name='cifar100'
)

print(f"\nCIFAR-100 GraSP Pruning Complete!")
print(f"Best Accuracy: {cifar100_best_acc:.2f}%")


TASK 1B: GRASP ITERATIVE PRUNING PIPELINE (ROBUST MODE)
Dataset: CIFAR100
Target Sparsity: 80.0%
Stage 1 (at 20% acc): 40.0% sparsity (50% of target)
Stage 2 (at 40% acc): 60.00000000000001% sparsity (75% of target)
Stage 3 (at 60% acc): 80.0% sparsity (100% of target)

CHECKING FOR EXISTING CHECKPOINTS (RECOVERY MODE)
  Results JSON:         ✗ Not found
  Final checkpoint:     ✗ Not found
  Stage 3 mask:         ✗ Not found
  Stage 3 checkpoint:   ✗ Not found
  Stage 2 mask:         ✗ Not found
  Stage 2 checkpoint:   ✗ Not found
  Stage 1 mask:         ✗ Not found
  Stage 1 checkpoint:   ✗ Not found

Initializing model with random weights...
✓ Model weights initialized with random values

STAGE 1 TRAINING: Target accuracy 20.0%

Fine-tuning to 20.0% accuracy (max 200 epochs)

Starting training...



Epoch [1/200] Loss: 5.1744 | LR: 0.010000 | Val Top-1: 2.10% | Val Top-5: 7.60%


Epoch [2/200] Loss: 4.5502 | LR: 0.010000 | Val Top-1: 3.48% | Val Top-5: 14.64%


Epoch [3/200] Loss: 4.4051 | LR: 0.010000 | Val Top-1: 6.68% | Val Top-5: 22.52%


Epoch [4/200] Loss: 4.2460 | LR: 0.010000 | Val Top-1: 8.29% | Val Top-5: 27.62%


Epoch [5/200] Loss: 4.0940 | LR: 0.010000 | Val Top-1: 10.87% | Val Top-5: 33.34%


Epoch [6/200] Loss: 3.9179 | LR: 0.010000 | Val Top-1: 14.54% | Val Top-5: 40.35%


Epoch [7/200] Loss: 3.7123 | LR: 0.010000 | Val Top-1: 15.96% | Val Top-5: 43.26%


Epoch [8/200] Loss: 3.4874 | LR: 0.010000 | Val Top-1: 20.82% | Val Top-5: 49.94%

✓ Target accuracy 20.0% reached! (Val Top-1: 20.82%)
✓ Final checkpoint saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints/cifar100_stage1_trained.pt
Fine-tuning completed. Best accuracy: 20.82%

STAGE 1 PRUNING: Target sparsity 40.0%

GraSP Pruning (Algorithm 1) - Target sparsity: 40.0%
Computing Hessian-gradient product (Algorithm 2)...
  Calibration batches: 4
✓ Hessian-gradient product computed over 4 batches
  Layers processed: 29

Computed threshold: 0.000036
Prunable parameters: 15,290,176
Parameters to keep: 9,174,105 (60.0% of prunable)
Parameters to prune: 6,116,071 (40.0% of prunable)
  features.0: 49.54% sparsity (856/1,728 pruned)
  features.1: 50.00% sparsity (32/64 pruned)
  features.3: 49.68% sparsity (18,314/36,864 pruned)
  features.4: 50.00% sparsity (32/64 pruned)
  features.7: 50.18% sparsity (36,999/73,728 pruned)
  features.8: 42.97% sparsity (55/128 pruned

Epoch [1/200] Loss: 5.8667 | LR: 0.010000 | Val Top-1: 1.50% | Val Top-5: 6.26%


Epoch [2/200] Loss: 4.6001 | LR: 0.010000 | Val Top-1: 1.32% | Val Top-5: 6.69%


Epoch [3/200] Loss: 4.5927 | LR: 0.010000 | Val Top-1: 1.31% | Val Top-5: 6.62%


Epoch [4/200] Loss: 4.5851 | LR: 0.010000 | Val Top-1: 1.50% | Val Top-5: 6.95%


Epoch [5/200] Loss: 4.5812 | LR: 0.010000 | Val Top-1: 1.47% | Val Top-5: 7.14%


Epoch [6/200] Loss: 4.5780 | LR: 0.010000 | Val Top-1: 1.55% | Val Top-5: 7.11%


Epoch [7/200] Loss: 4.5741 | LR: 0.010000 | Val Top-1: 1.46% | Val Top-5: 7.39%


Epoch [8/200] Loss: 4.5684 | LR: 0.010000 | Val Top-1: 1.57% | Val Top-5: 7.13%


Epoch [9/200] Loss: 4.5671 | LR: 0.010000 | Val Top-1: 1.45% | Val Top-5: 6.98%


Epoch [10/200] Loss: 4.5658 | LR: 0.010000 | Val Top-1: 1.55% | Val Top-5: 7.01%


Epoch [11/200] Loss: 4.5624 | LR: 0.010000 | Val Top-1: 1.69% | Val Top-5: 7.06%


Epoch [12/200] Loss: 4.5608 | LR: 0.010000 | Val Top-1: 1.58% | Val Top-5: 6.81%


Epoch [13/200] Loss: 4.5551 | LR: 0.010000 | Val Top-1: 1.58% | Val Top-5: 7.02%


Epoch [14/200] Loss: 4.5539 | LR: 0.010000 | Val Top-1: 1.81% | Val Top-5: 7.49%


Epoch [15/200] Loss: 4.5513 | LR: 0.010000 | Val Top-1: 1.80% | Val Top-5: 7.73%


Epoch [16/200] Loss: 4.5468 | LR: 0.010000 | Val Top-1: 1.55% | Val Top-5: 7.37%


Epoch [17/200] Loss: 4.5427 | LR: 0.010000 | Val Top-1: 1.66% | Val Top-5: 7.90%


Epoch [18/200] Loss: 4.5227 | LR: 0.010000 | Val Top-1: 2.17% | Val Top-5: 8.62%


Epoch [19/200] Loss: 4.4757 | LR: 0.010000 | Val Top-1: 2.14% | Val Top-5: 10.16%


Epoch [20/200] Loss: 4.4168 | LR: 0.010000 | Val Top-1: 2.39% | Val Top-5: 11.68%


Epoch [21/200] Loss: 4.3753 | LR: 0.010000 | Val Top-1: 2.82% | Val Top-5: 13.14%


Epoch [22/200] Loss: 4.3190 | LR: 0.010000 | Val Top-1: 3.18% | Val Top-5: 15.10%


Epoch [23/200] Loss: 4.2757 | LR: 0.010000 | Val Top-1: 3.69% | Val Top-5: 16.20%


Epoch [24/200] Loss: 4.2436 | LR: 0.010000 | Val Top-1: 3.81% | Val Top-5: 17.15%


Epoch [25/200] Loss: 4.2197 | LR: 0.010000 | Val Top-1: 3.86% | Val Top-5: 17.82%


Epoch [26/200] Loss: 4.1998 | LR: 0.010000 | Val Top-1: 4.37% | Val Top-5: 18.46%


Epoch [27/200] Loss: 4.1837 | LR: 0.010000 | Val Top-1: 4.01% | Val Top-5: 18.70%


Epoch [28/200] Loss: 4.1555 | LR: 0.010000 | Val Top-1: 4.79% | Val Top-5: 20.32%


Epoch [29/200] Loss: 4.1091 | LR: 0.010000 | Val Top-1: 5.57% | Val Top-5: 22.56%


Epoch [30/200] Loss: 4.0572 | LR: 0.010000 | Val Top-1: 5.65% | Val Top-5: 23.29%


Epoch [31/200] Loss: 4.0176 | LR: 0.010000 | Val Top-1: 6.46% | Val Top-5: 25.25%


Epoch [32/200] Loss: 3.9758 | LR: 0.010000 | Val Top-1: 6.86% | Val Top-5: 26.91%


Epoch [33/200] Loss: 3.9308 | LR: 0.010000 | Val Top-1: 7.57% | Val Top-5: 27.44%


Epoch [34/200] Loss: 3.8843 | LR: 0.010000 | Val Top-1: 7.39% | Val Top-5: 27.88%


Epoch [35/200] Loss: 3.8370 | LR: 0.010000 | Val Top-1: 8.97% | Val Top-5: 32.20%


Epoch [36/200] Loss: 3.7722 | LR: 0.010000 | Val Top-1: 9.53% | Val Top-5: 34.04%


Epoch [37/200] Loss: 3.7022 | LR: 0.010000 | Val Top-1: 10.98% | Val Top-5: 35.94%


Epoch [38/200] Loss: 3.6170 | LR: 0.010000 | Val Top-1: 13.33% | Val Top-5: 40.25%


Epoch [39/200] Loss: 3.5522 | LR: 0.010000 | Val Top-1: 11.68% | Val Top-5: 39.95%


Epoch [40/200] Loss: 3.4955 | LR: 0.010000 | Val Top-1: 13.39% | Val Top-5: 41.09%


Epoch [41/200] Loss: 3.4207 | LR: 0.010000 | Val Top-1: 15.58% | Val Top-5: 45.75%


Epoch [42/200] Loss: 3.3361 | LR: 0.010000 | Val Top-1: 16.94% | Val Top-5: 47.87%


Epoch [43/200] Loss: 3.2613 | LR: 0.010000 | Val Top-1: 17.03% | Val Top-5: 49.28%


Epoch [44/200] Loss: 3.1973 | LR: 0.010000 | Val Top-1: 19.10% | Val Top-5: 51.77%


Epoch [45/200] Loss: 3.1287 | LR: 0.010000 | Val Top-1: 18.10% | Val Top-5: 51.53%


Epoch [46/200] Loss: 3.0704 | LR: 0.010000 | Val Top-1: 20.27% | Val Top-5: 52.39%


Epoch [47/200] Loss: 3.0070 | LR: 0.010000 | Val Top-1: 21.34% | Val Top-5: 53.43%


Epoch [48/200] Loss: 2.9576 | LR: 0.010000 | Val Top-1: 23.04% | Val Top-5: 57.87%


Epoch [49/200] Loss: 2.8913 | LR: 0.010000 | Val Top-1: 25.33% | Val Top-5: 59.17%


Epoch [50/200] Loss: 2.8410 | LR: 0.010000 | Val Top-1: 26.40% | Val Top-5: 60.94%


Epoch [51/200] Loss: 2.7690 | LR: 0.010000 | Val Top-1: 24.85% | Val Top-5: 57.90%


Epoch [52/200] Loss: 2.7284 | LR: 0.010000 | Val Top-1: 25.99% | Val Top-5: 61.80%


Epoch [53/200] Loss: 2.6639 | LR: 0.010000 | Val Top-1: 30.57% | Val Top-5: 64.96%


Epoch [54/200] Loss: 2.6046 | LR: 0.010000 | Val Top-1: 31.88% | Val Top-5: 66.28%


Epoch [55/200] Loss: 2.5500 | LR: 0.010000 | Val Top-1: 33.86% | Val Top-5: 69.60%


Epoch [56/200] Loss: 2.4941 | LR: 0.010000 | Val Top-1: 32.47% | Val Top-5: 66.34%


Epoch [57/200] Loss: 2.4468 | LR: 0.010000 | Val Top-1: 35.21% | Val Top-5: 69.34%


Epoch [58/200] Loss: 2.3796 | LR: 0.010000 | Val Top-1: 35.47% | Val Top-5: 70.52%


Epoch [59/200] Loss: 2.3328 | LR: 0.010000 | Val Top-1: 33.36% | Val Top-5: 67.03%


Epoch [60/200] Loss: 2.2853 | LR: 0.002000 | Val Top-1: 39.04% | Val Top-5: 72.64%


Epoch [61/200] Loss: 2.0575 | LR: 0.002000 | Val Top-1: 45.47% | Val Top-5: 77.61%

✓ Target accuracy 40.0% reached! (Val Top-1: 45.47%)
✓ Final checkpoint saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints/cifar100_stage2_trained.pt
Fine-tuning completed. Best accuracy: 45.47%

STAGE 2 PRUNING: Target sparsity 60.00000000000001%

GraSP Pruning (Algorithm 1) - Target sparsity: 60.00000000000001%
Computing Hessian-gradient product (Algorithm 2)...
  Calibration batches: 4
✓ Hessian-gradient product computed over 4 batches
  Layers processed: 29

Computed threshold: -0.000007
Prunable parameters: 9,174,104
Parameters to keep: 3,669,641 (40.0% of prunable)
Parameters to prune: 5,504,463 (60.0% of prunable)
  features.0: 73.32% sparsity (1,267/1,728 pruned)
  features.1: 67.19% sparsity (43/64 pruned)
  features.3: 74.80% sparsity (27,573/36,864 pruned)
  features.4: 67.19% sparsity (43/64 pruned)
  features.7: 74.73% sparsity (55,094/73,728 pruned)
  features.8: 6

Epoch [1/200] Loss: 4.6538 | LR: 0.010000 | Val Top-1: 1.44% | Val Top-5: 6.42%


Epoch [2/200] Loss: 4.5798 | LR: 0.010000 | Val Top-1: 1.58% | Val Top-5: 7.09%


Epoch [3/200] Loss: 4.5691 | LR: 0.010000 | Val Top-1: 1.63% | Val Top-5: 7.34%


Epoch [4/200] Loss: 4.5240 | LR: 0.010000 | Val Top-1: 2.60% | Val Top-5: 10.76%


Epoch [5/200] Loss: 4.4353 | LR: 0.010000 | Val Top-1: 2.71% | Val Top-5: 11.61%


Epoch [6/200] Loss: 4.3810 | LR: 0.010000 | Val Top-1: 2.89% | Val Top-5: 11.76%


Epoch [7/200] Loss: 4.3231 | LR: 0.010000 | Val Top-1: 3.63% | Val Top-5: 15.25%


Epoch [8/200] Loss: 4.2511 | LR: 0.010000 | Val Top-1: 3.55% | Val Top-5: 16.30%


Epoch [9/200] Loss: 4.2054 | LR: 0.010000 | Val Top-1: 4.24% | Val Top-5: 17.97%


Epoch [10/200] Loss: 4.1691 | LR: 0.010000 | Val Top-1: 4.65% | Val Top-5: 19.83%


Epoch [11/200] Loss: 4.1391 | LR: 0.010000 | Val Top-1: 5.16% | Val Top-5: 20.99%


Epoch [12/200] Loss: 4.0895 | LR: 0.010000 | Val Top-1: 5.89% | Val Top-5: 24.52%


Epoch [13/200] Loss: 3.9909 | LR: 0.010000 | Val Top-1: 8.49% | Val Top-5: 29.28%


Epoch [14/200] Loss: 3.8893 | LR: 0.010000 | Val Top-1: 8.72% | Val Top-5: 29.89%


Epoch [15/200] Loss: 3.8029 | LR: 0.010000 | Val Top-1: 10.45% | Val Top-5: 35.21%


Epoch [16/200] Loss: 3.7153 | LR: 0.010000 | Val Top-1: 11.94% | Val Top-5: 37.49%


Epoch [17/200] Loss: 3.6378 | LR: 0.010000 | Val Top-1: 12.98% | Val Top-5: 40.82%


Epoch [18/200] Loss: 3.5630 | LR: 0.010000 | Val Top-1: 13.09% | Val Top-5: 41.73%


Epoch [19/200] Loss: 3.4900 | LR: 0.010000 | Val Top-1: 14.61% | Val Top-5: 42.91%


Epoch [20/200] Loss: 3.4112 | LR: 0.010000 | Val Top-1: 17.16% | Val Top-5: 48.81%


Epoch [21/200] Loss: 3.3093 | LR: 0.010000 | Val Top-1: 17.59% | Val Top-5: 50.15%


Epoch [22/200] Loss: 3.2315 | LR: 0.010000 | Val Top-1: 19.18% | Val Top-5: 53.00%


Epoch [23/200] Loss: 3.1449 | LR: 0.010000 | Val Top-1: 20.50% | Val Top-5: 55.05%


Epoch [24/200] Loss: 3.0832 | LR: 0.010000 | Val Top-1: 23.80% | Val Top-5: 58.57%


Epoch [25/200] Loss: 3.0105 | LR: 0.010000 | Val Top-1: 23.88% | Val Top-5: 60.41%


Epoch [26/200] Loss: 2.9498 | LR: 0.010000 | Val Top-1: 24.76% | Val Top-5: 59.51%


Epoch [27/200] Loss: 2.8979 | LR: 0.010000 | Val Top-1: 26.16% | Val Top-5: 61.50%


Epoch [28/200] Loss: 2.8404 | LR: 0.010000 | Val Top-1: 28.17% | Val Top-5: 62.36%


Epoch [29/200] Loss: 2.7796 | LR: 0.010000 | Val Top-1: 31.04% | Val Top-5: 65.35%


Epoch [30/200] Loss: 2.7292 | LR: 0.010000 | Val Top-1: 30.22% | Val Top-5: 64.86%


Epoch [31/200] Loss: 2.6847 | LR: 0.010000 | Val Top-1: 31.73% | Val Top-5: 66.13%


Epoch [32/200] Loss: 2.6397 | LR: 0.010000 | Val Top-1: 33.92% | Val Top-5: 67.75%


Epoch [33/200] Loss: 2.6075 | LR: 0.010000 | Val Top-1: 29.86% | Val Top-5: 64.30%


Epoch [34/200] Loss: 2.5541 | LR: 0.010000 | Val Top-1: 35.23% | Val Top-5: 69.62%


Epoch [35/200] Loss: 2.5141 | LR: 0.010000 | Val Top-1: 33.87% | Val Top-5: 69.22%


Epoch [36/200] Loss: 2.4798 | LR: 0.010000 | Val Top-1: 36.68% | Val Top-5: 69.29%


Epoch [37/200] Loss: 2.4459 | LR: 0.010000 | Val Top-1: 36.18% | Val Top-5: 69.91%


Epoch [38/200] Loss: 2.3986 | LR: 0.010000 | Val Top-1: 37.18% | Val Top-5: 70.85%


Epoch [39/200] Loss: 2.3721 | LR: 0.010000 | Val Top-1: 38.94% | Val Top-5: 72.62%


Epoch [40/200] Loss: 2.3405 | LR: 0.010000 | Val Top-1: 39.68% | Val Top-5: 73.17%


Epoch [41/200] Loss: 2.3131 | LR: 0.010000 | Val Top-1: 39.59% | Val Top-5: 72.19%


Epoch [42/200] Loss: 2.2774 | LR: 0.010000 | Val Top-1: 40.38% | Val Top-5: 73.43%


Epoch [43/200] Loss: 2.2502 | LR: 0.010000 | Val Top-1: 40.18% | Val Top-5: 73.36%


Epoch [44/200] Loss: 2.2272 | LR: 0.010000 | Val Top-1: 39.12% | Val Top-5: 71.53%


Epoch [45/200] Loss: 2.1908 | LR: 0.010000 | Val Top-1: 40.62% | Val Top-5: 72.49%


Epoch [46/200] Loss: 2.1647 | LR: 0.010000 | Val Top-1: 42.11% | Val Top-5: 73.64%


Epoch [47/200] Loss: 2.1480 | LR: 0.010000 | Val Top-1: 41.87% | Val Top-5: 73.46%


Epoch [48/200] Loss: 2.1175 | LR: 0.010000 | Val Top-1: 43.30% | Val Top-5: 75.05%


Epoch [49/200] Loss: 2.0920 | LR: 0.010000 | Val Top-1: 42.89% | Val Top-5: 74.31%


Epoch [50/200] Loss: 2.0682 | LR: 0.010000 | Val Top-1: 45.19% | Val Top-5: 75.65%


Epoch [51/200] Loss: 2.0424 | LR: 0.010000 | Val Top-1: 44.59% | Val Top-5: 74.97%


Epoch [52/200] Loss: 2.0229 | LR: 0.010000 | Val Top-1: 45.59% | Val Top-5: 75.92%


Epoch [53/200] Loss: 2.0051 | LR: 0.010000 | Val Top-1: 47.72% | Val Top-5: 77.81%


Epoch [54/200] Loss: 1.9672 | LR: 0.010000 | Val Top-1: 44.15% | Val Top-5: 73.85%


Epoch [55/200] Loss: 1.9592 | LR: 0.010000 | Val Top-1: 47.43% | Val Top-5: 78.20%


Epoch [56/200] Loss: 1.9388 | LR: 0.010000 | Val Top-1: 48.10% | Val Top-5: 77.65%


Epoch [57/200] Loss: 1.9196 | LR: 0.010000 | Val Top-1: 48.26% | Val Top-5: 77.53%


Epoch [58/200] Loss: 1.9002 | LR: 0.010000 | Val Top-1: 48.61% | Val Top-5: 76.93%


Epoch [59/200] Loss: 1.8782 | LR: 0.010000 | Val Top-1: 48.29% | Val Top-5: 78.28%


Epoch [60/200] Loss: 1.8573 | LR: 0.002000 | Val Top-1: 49.55% | Val Top-5: 77.60%


Epoch [61/200] Loss: 1.6534 | LR: 0.002000 | Val Top-1: 55.93% | Val Top-5: 82.63%


Epoch [62/200] Loss: 1.5706 | LR: 0.002000 | Val Top-1: 56.04% | Val Top-5: 82.46%


Epoch [63/200] Loss: 1.5348 | LR: 0.002000 | Val Top-1: 55.95% | Val Top-5: 82.28%


Epoch [64/200] Loss: 1.5039 | LR: 0.002000 | Val Top-1: 56.30% | Val Top-5: 82.47%


Epoch [65/200] Loss: 1.4786 | LR: 0.002000 | Val Top-1: 56.55% | Val Top-5: 82.48%


Epoch [66/200] Loss: 1.4625 | LR: 0.002000 | Val Top-1: 56.83% | Val Top-5: 82.55%


Epoch [67/200] Loss: 1.4361 | LR: 0.002000 | Val Top-1: 56.92% | Val Top-5: 82.49%


Epoch [68/200] Loss: 1.4317 | LR: 0.002000 | Val Top-1: 56.81% | Val Top-5: 82.72%


Epoch [69/200] Loss: 1.4098 | LR: 0.002000 | Val Top-1: 57.41% | Val Top-5: 82.66%


Epoch [70/200] Loss: 1.3896 | LR: 0.002000 | Val Top-1: 57.23% | Val Top-5: 82.37%


Epoch [71/200] Loss: 1.3720 | LR: 0.002000 | Val Top-1: 57.47% | Val Top-5: 82.48%


Epoch [72/200] Loss: 1.3566 | LR: 0.002000 | Val Top-1: 57.50% | Val Top-5: 82.79%


Epoch [73/200] Loss: 1.3530 | LR: 0.002000 | Val Top-1: 58.05% | Val Top-5: 82.90%


Epoch [74/200] Loss: 1.3306 | LR: 0.002000 | Val Top-1: 58.04% | Val Top-5: 82.60%


Epoch [75/200] Loss: 1.3277 | LR: 0.002000 | Val Top-1: 57.87% | Val Top-5: 82.58%


Epoch [76/200] Loss: 1.3102 | LR: 0.002000 | Val Top-1: 57.75% | Val Top-5: 82.74%


Epoch [77/200] Loss: 1.3057 | LR: 0.002000 | Val Top-1: 58.14% | Val Top-5: 82.73%


Epoch [78/200] Loss: 1.2809 | LR: 0.002000 | Val Top-1: 57.63% | Val Top-5: 82.69%


Epoch [79/200] Loss: 1.2746 | LR: 0.002000 | Val Top-1: 57.70% | Val Top-5: 82.69%


Epoch [80/200] Loss: 1.2619 | LR: 0.002000 | Val Top-1: 57.84% | Val Top-5: 82.94%


Epoch [81/200] Loss: 1.2491 | LR: 0.002000 | Val Top-1: 57.94% | Val Top-5: 82.57%


Epoch [82/200] Loss: 1.2368 | LR: 0.002000 | Val Top-1: 58.57% | Val Top-5: 83.31%


Epoch [83/200] Loss: 1.2276 | LR: 0.002000 | Val Top-1: 58.15% | Val Top-5: 82.78%


Epoch [84/200] Loss: 1.2208 | LR: 0.002000 | Val Top-1: 58.59% | Val Top-5: 83.49%


Epoch [85/200] Loss: 1.2076 | LR: 0.002000 | Val Top-1: 58.47% | Val Top-5: 82.83%


Epoch [86/200] Loss: 1.2110 | LR: 0.002000 | Val Top-1: 58.59% | Val Top-5: 83.18%


Epoch [87/200] Loss: 1.1914 | LR: 0.002000 | Val Top-1: 58.66% | Val Top-5: 83.10%


Epoch [88/200] Loss: 1.1752 | LR: 0.002000 | Val Top-1: 58.39% | Val Top-5: 82.95%


Epoch [89/200] Loss: 1.1754 | LR: 0.002000 | Val Top-1: 59.18% | Val Top-5: 82.84%


Epoch [90/200] Loss: 1.1657 | LR: 0.002000 | Val Top-1: 59.04% | Val Top-5: 82.68%


Epoch [91/200] Loss: 1.1559 | LR: 0.002000 | Val Top-1: 59.00% | Val Top-5: 82.69%


Epoch [92/200] Loss: 1.1375 | LR: 0.002000 | Val Top-1: 59.12% | Val Top-5: 82.48%


Epoch [93/200] Loss: 1.1334 | LR: 0.002000 | Val Top-1: 58.99% | Val Top-5: 83.03%


Epoch [94/200] Loss: 1.1199 | LR: 0.002000 | Val Top-1: 59.27% | Val Top-5: 82.71%


Epoch [95/200] Loss: 1.1171 | LR: 0.002000 | Val Top-1: 59.08% | Val Top-5: 82.76%


Epoch [96/200] Loss: 1.1077 | LR: 0.002000 | Val Top-1: 59.10% | Val Top-5: 82.95%


Epoch [97/200] Loss: 1.0968 | LR: 0.002000 | Val Top-1: 58.81% | Val Top-5: 82.74%


Epoch [98/200] Loss: 1.0917 | LR: 0.002000 | Val Top-1: 59.45% | Val Top-5: 82.79%


Epoch [99/200] Loss: 1.0845 | LR: 0.002000 | Val Top-1: 59.33% | Val Top-5: 82.55%


Epoch [100/200] Loss: 1.0808 | LR: 0.002000 | Val Top-1: 59.28% | Val Top-5: 82.81%


Epoch [101/200] Loss: 1.0624 | LR: 0.002000 | Val Top-1: 59.08% | Val Top-5: 82.66%


Epoch [102/200] Loss: 1.0553 | LR: 0.002000 | Val Top-1: 59.17% | Val Top-5: 82.53%


Epoch [103/200] Loss: 1.0624 | LR: 0.002000 | Val Top-1: 59.28% | Val Top-5: 82.40%


Epoch [104/200] Loss: 1.0468 | LR: 0.002000 | Val Top-1: 59.59% | Val Top-5: 82.45%


Epoch [105/200] Loss: 1.0292 | LR: 0.002000 | Val Top-1: 59.45% | Val Top-5: 83.01%


Epoch [106/200] Loss: 1.0255 | LR: 0.002000 | Val Top-1: 59.33% | Val Top-5: 83.18%


Epoch [107/200] Loss: 1.0315 | LR: 0.002000 | Val Top-1: 59.47% | Val Top-5: 82.57%


Epoch [108/200] Loss: 1.0189 | LR: 0.002000 | Val Top-1: 59.35% | Val Top-5: 82.46%


Epoch [109/200] Loss: 1.0048 | LR: 0.002000 | Val Top-1: 58.57% | Val Top-5: 82.43%


Epoch [110/200] Loss: 1.0058 | LR: 0.002000 | Val Top-1: 59.72% | Val Top-5: 82.54%


Epoch [111/200] Loss: 0.9942 | LR: 0.002000 | Val Top-1: 59.52% | Val Top-5: 82.63%


Epoch [112/200] Loss: 0.9941 | LR: 0.002000 | Val Top-1: 59.81% | Val Top-5: 82.68%


Epoch [113/200] Loss: 0.9842 | LR: 0.002000 | Val Top-1: 60.24% | Val Top-5: 82.65%

✓ Target accuracy 60.0% reached! (Val Top-1: 60.24%)
✓ Final checkpoint saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/checkpoints/cifar100_stage3_trained.pt
Fine-tuning completed. Best accuracy: 60.24%

STAGE 3 PRUNING: Target sparsity 80.0%

GraSP Pruning (Algorithm 1) - Target sparsity: 80.0%
Computing Hessian-gradient product (Algorithm 2)...
  Calibration batches: 4
✓ Hessian-gradient product computed over 4 batches
  Layers processed: 29

Computed threshold: -0.020301
Prunable parameters: 3,669,640
Parameters to keep: 733,927 (20.0% of prunable)
Parameters to prune: 2,935,713 (80.0% of prunable)
  features.0: 88.83% sparsity (1,535/1,728 pruned)
  features.1: 87.50% sparsity (56/64 pruned)
  features.3: 89.14% sparsity (32,859/36,864 pruned)
  features.4: 84.38% sparsity (54/64 pruned)
  features.7: 88.44% sparsity (65,203/73,728 pruned)
  features.8: 83.59% sparsity (107/128 pru

Epoch [1/200] Loss: 4.6815 | LR: 0.010000 | Val Top-1: 1.82% | Val Top-5: 6.90%


Epoch [2/200] Loss: 4.5485 | LR: 0.010000 | Val Top-1: 2.37% | Val Top-5: 10.02%


Epoch [3/200] Loss: 4.4046 | LR: 0.010000 | Val Top-1: 3.81% | Val Top-5: 18.19%


Epoch [4/200] Loss: 4.2061 | LR: 0.010000 | Val Top-1: 6.62% | Val Top-5: 26.41%


Epoch [5/200] Loss: 4.0174 | LR: 0.010000 | Val Top-1: 8.31% | Val Top-5: 29.38%


Epoch [6/200] Loss: 3.8830 | LR: 0.010000 | Val Top-1: 10.95% | Val Top-5: 35.34%


Epoch [7/200] Loss: 3.7852 | LR: 0.010000 | Val Top-1: 13.19% | Val Top-5: 41.05%


Epoch [8/200] Loss: 3.7007 | LR: 0.010000 | Val Top-1: 15.00% | Val Top-5: 42.77%


Epoch [9/200] Loss: 3.6054 | LR: 0.010000 | Val Top-1: 16.63% | Val Top-5: 47.67%


Epoch [10/200] Loss: 3.5274 | LR: 0.010000 | Val Top-1: 17.00% | Val Top-5: 49.54%


Epoch [11/200] Loss: 3.4649 | LR: 0.010000 | Val Top-1: 19.66% | Val Top-5: 53.06%


Epoch [12/200] Loss: 3.3939 | LR: 0.010000 | Val Top-1: 20.90% | Val Top-5: 54.96%


Epoch [13/200] Loss: 3.3339 | LR: 0.010000 | Val Top-1: 20.73% | Val Top-5: 53.49%


Epoch [14/200] Loss: 3.2772 | LR: 0.010000 | Val Top-1: 23.76% | Val Top-5: 57.92%


Epoch [15/200] Loss: 3.2345 | LR: 0.010000 | Val Top-1: 24.20% | Val Top-5: 58.11%


Epoch [16/200] Loss: 3.1822 | LR: 0.010000 | Val Top-1: 25.72% | Val Top-5: 58.23%


Epoch [17/200] Loss: 3.1437 | LR: 0.010000 | Val Top-1: 27.29% | Val Top-5: 59.95%


Epoch [18/200] Loss: 3.0980 | LR: 0.010000 | Val Top-1: 27.93% | Val Top-5: 61.13%


Epoch [19/200] Loss: 3.0500 | LR: 0.010000 | Val Top-1: 28.48% | Val Top-5: 62.85%


Epoch [20/200] Loss: 3.0103 | LR: 0.010000 | Val Top-1: 29.57% | Val Top-5: 62.41%


Epoch [21/200] Loss: 2.9801 | LR: 0.010000 | Val Top-1: 30.99% | Val Top-5: 63.67%


Epoch [22/200] Loss: 2.9465 | LR: 0.010000 | Val Top-1: 31.68% | Val Top-5: 64.48%


Epoch [23/200] Loss: 2.9089 | LR: 0.010000 | Val Top-1: 32.65% | Val Top-5: 65.66%


Epoch [24/200] Loss: 2.8749 | LR: 0.010000 | Val Top-1: 33.21% | Val Top-5: 64.70%


Epoch [25/200] Loss: 2.8404 | LR: 0.010000 | Val Top-1: 34.49% | Val Top-5: 67.06%


Epoch [26/200] Loss: 2.8161 | LR: 0.010000 | Val Top-1: 34.19% | Val Top-5: 65.85%


Epoch [27/200] Loss: 2.7827 | LR: 0.010000 | Val Top-1: 35.70% | Val Top-5: 68.39%


Epoch [28/200] Loss: 2.7579 | LR: 0.010000 | Val Top-1: 35.41% | Val Top-5: 67.33%


Epoch [29/200] Loss: 2.7303 | LR: 0.010000 | Val Top-1: 36.76% | Val Top-5: 67.90%


Epoch [30/200] Loss: 2.6923 | LR: 0.010000 | Val Top-1: 37.47% | Val Top-5: 69.49%


Epoch [31/200] Loss: 2.6747 | LR: 0.010000 | Val Top-1: 37.70% | Val Top-5: 69.95%


Epoch [32/200] Loss: 2.6460 | LR: 0.010000 | Val Top-1: 38.28% | Val Top-5: 69.86%


Epoch [33/200] Loss: 2.6198 | LR: 0.010000 | Val Top-1: 39.04% | Val Top-5: 71.11%


Epoch [34/200] Loss: 2.5952 | LR: 0.010000 | Val Top-1: 38.46% | Val Top-5: 70.81%


Epoch [35/200] Loss: 2.5723 | LR: 0.010000 | Val Top-1: 40.14% | Val Top-5: 71.55%


Epoch [36/200] Loss: 2.5550 | LR: 0.010000 | Val Top-1: 39.50% | Val Top-5: 70.72%


Epoch [37/200] Loss: 2.5315 | LR: 0.010000 | Val Top-1: 40.27% | Val Top-5: 71.52%


Epoch [38/200] Loss: 2.5127 | LR: 0.010000 | Val Top-1: 41.68% | Val Top-5: 73.44%


Epoch [39/200] Loss: 2.4847 | LR: 0.010000 | Val Top-1: 41.07% | Val Top-5: 73.14%


Epoch [40/200] Loss: 2.4681 | LR: 0.010000 | Val Top-1: 44.13% | Val Top-5: 75.24%


Epoch [41/200] Loss: 2.4562 | LR: 0.010000 | Val Top-1: 44.74% | Val Top-5: 74.48%


Epoch [42/200] Loss: 2.4313 | LR: 0.010000 | Val Top-1: 44.40% | Val Top-5: 74.66%


Epoch [43/200] Loss: 2.4197 | LR: 0.010000 | Val Top-1: 44.04% | Val Top-5: 74.35%


Epoch [44/200] Loss: 2.3896 | LR: 0.010000 | Val Top-1: 43.77% | Val Top-5: 74.02%


Epoch [45/200] Loss: 2.3839 | LR: 0.010000 | Val Top-1: 44.33% | Val Top-5: 73.49%


Epoch [46/200] Loss: 2.3645 | LR: 0.010000 | Val Top-1: 43.48% | Val Top-5: 72.04%


Epoch [47/200] Loss: 2.3503 | LR: 0.010000 | Val Top-1: 45.73% | Val Top-5: 74.39%


Epoch [48/200] Loss: 2.3365 | LR: 0.010000 | Val Top-1: 45.83% | Val Top-5: 75.01%


Epoch [49/200] Loss: 2.3113 | LR: 0.010000 | Val Top-1: 45.85% | Val Top-5: 74.36%


Epoch [50/200] Loss: 2.3030 | LR: 0.010000 | Val Top-1: 45.82% | Val Top-5: 74.89%


Epoch [51/200] Loss: 2.2903 | LR: 0.010000 | Val Top-1: 46.26% | Val Top-5: 75.58%


Epoch [52/200] Loss: 2.2754 | LR: 0.010000 | Val Top-1: 47.59% | Val Top-5: 76.85%


Epoch [53/200] Loss: 2.2624 | LR: 0.010000 | Val Top-1: 46.60% | Val Top-5: 76.25%


Epoch [54/200] Loss: 2.2559 | LR: 0.010000 | Val Top-1: 48.42% | Val Top-5: 76.97%


Epoch [55/200] Loss: 2.2412 | LR: 0.010000 | Val Top-1: 47.71% | Val Top-5: 75.99%


Epoch [56/200] Loss: 2.2251 | LR: 0.010000 | Val Top-1: 47.34% | Val Top-5: 75.56%


Epoch [57/200] Loss: 2.2161 | LR: 0.010000 | Val Top-1: 48.18% | Val Top-5: 76.52%


Epoch [58/200] Loss: 2.1994 | LR: 0.010000 | Val Top-1: 47.31% | Val Top-5: 76.09%


Epoch [59/200] Loss: 2.1880 | LR: 0.010000 | Val Top-1: 49.66% | Val Top-5: 77.64%


Epoch [60/200] Loss: 2.1715 | LR: 0.002000 | Val Top-1: 48.80% | Val Top-5: 76.74%


Epoch [61/200] Loss: 2.0073 | LR: 0.002000 | Val Top-1: 53.92% | Val Top-5: 80.71%


Epoch [62/200] Loss: 1.9493 | LR: 0.002000 | Val Top-1: 54.61% | Val Top-5: 80.87%


Epoch [63/200] Loss: 1.9184 | LR: 0.002000 | Val Top-1: 54.68% | Val Top-5: 81.25%


Epoch [64/200] Loss: 1.8997 | LR: 0.002000 | Val Top-1: 54.84% | Val Top-5: 81.26%


Epoch [65/200] Loss: 1.8735 | LR: 0.002000 | Val Top-1: 54.83% | Val Top-5: 81.27%


Epoch [66/200] Loss: 1.8697 | LR: 0.002000 | Val Top-1: 54.94% | Val Top-5: 81.77%


Epoch [67/200] Loss: 1.8568 | LR: 0.002000 | Val Top-1: 55.05% | Val Top-5: 81.11%


Epoch [68/200] Loss: 1.8337 | LR: 0.002000 | Val Top-1: 55.47% | Val Top-5: 81.25%


Epoch [69/200] Loss: 1.8262 | LR: 0.002000 | Val Top-1: 55.54% | Val Top-5: 81.32%


Epoch [70/200] Loss: 1.8149 | LR: 0.002000 | Val Top-1: 55.73% | Val Top-5: 81.84%


Epoch [71/200] Loss: 1.8055 | LR: 0.002000 | Val Top-1: 55.62% | Val Top-5: 81.07%


Epoch [72/200] Loss: 1.7943 | LR: 0.002000 | Val Top-1: 56.09% | Val Top-5: 81.96%


Epoch [73/200] Loss: 1.7831 | LR: 0.002000 | Val Top-1: 56.26% | Val Top-5: 81.70%


Epoch [74/200] Loss: 1.7741 | LR: 0.002000 | Val Top-1: 56.32% | Val Top-5: 81.44%


Epoch [75/200] Loss: 1.7673 | LR: 0.002000 | Val Top-1: 56.76% | Val Top-5: 81.43%


Epoch [76/200] Loss: 1.7593 | LR: 0.002000 | Val Top-1: 55.93% | Val Top-5: 81.68%


Epoch [77/200] Loss: 1.7461 | LR: 0.002000 | Val Top-1: 56.20% | Val Top-5: 81.82%


Epoch [78/200] Loss: 1.7455 | LR: 0.002000 | Val Top-1: 56.49% | Val Top-5: 81.90%


Epoch [79/200] Loss: 1.7400 | LR: 0.002000 | Val Top-1: 56.81% | Val Top-5: 81.94%


Epoch [80/200] Loss: 1.7214 | LR: 0.002000 | Val Top-1: 56.95% | Val Top-5: 81.97%


Epoch [81/200] Loss: 1.7151 | LR: 0.002000 | Val Top-1: 56.39% | Val Top-5: 81.81%


Epoch [82/200] Loss: 1.7202 | LR: 0.002000 | Val Top-1: 56.77% | Val Top-5: 81.80%


Epoch [83/200] Loss: 1.7133 | LR: 0.002000 | Val Top-1: 57.03% | Val Top-5: 82.17%


Epoch [84/200] Loss: 1.7006 | LR: 0.002000 | Val Top-1: 56.57% | Val Top-5: 81.61%


Epoch [85/200] Loss: 1.6950 | LR: 0.002000 | Val Top-1: 56.58% | Val Top-5: 81.72%


Epoch [86/200] Loss: 1.6918 | LR: 0.002000 | Val Top-1: 56.86% | Val Top-5: 81.61%


Epoch [87/200] Loss: 1.6809 | LR: 0.002000 | Val Top-1: 56.69% | Val Top-5: 81.42%


Epoch [88/200] Loss: 1.6813 | LR: 0.002000 | Val Top-1: 57.52% | Val Top-5: 82.09%


Epoch [89/200] Loss: 1.6703 | LR: 0.002000 | Val Top-1: 56.77% | Val Top-5: 81.88%


Epoch [90/200] Loss: 1.6765 | LR: 0.002000 | Val Top-1: 57.20% | Val Top-5: 81.99%


Epoch [91/200] Loss: 1.6606 | LR: 0.002000 | Val Top-1: 57.01% | Val Top-5: 82.05%


Epoch [92/200] Loss: 1.6579 | LR: 0.002000 | Val Top-1: 57.30% | Val Top-5: 81.78%


Epoch [93/200] Loss: 1.6573 | LR: 0.002000 | Val Top-1: 56.93% | Val Top-5: 81.98%


Epoch [94/200] Loss: 1.6427 | LR: 0.002000 | Val Top-1: 57.30% | Val Top-5: 81.75%


Epoch [95/200] Loss: 1.6405 | LR: 0.002000 | Val Top-1: 57.23% | Val Top-5: 82.23%


Epoch [96/200] Loss: 1.6380 | LR: 0.002000 | Val Top-1: 57.26% | Val Top-5: 82.18%


Epoch [97/200] Loss: 1.6304 | LR: 0.002000 | Val Top-1: 57.37% | Val Top-5: 82.19%


Epoch [98/200] Loss: 1.6284 | LR: 0.002000 | Val Top-1: 57.40% | Val Top-5: 82.06%


Epoch [99/200] Loss: 1.6260 | LR: 0.002000 | Val Top-1: 57.49% | Val Top-5: 82.16%


Epoch [100/200] Loss: 1.6165 | LR: 0.002000 | Val Top-1: 57.43% | Val Top-5: 81.80%


Epoch [101/200] Loss: 1.6109 | LR: 0.002000 | Val Top-1: 57.28% | Val Top-5: 81.87%


Epoch [102/200] Loss: 1.6064 | LR: 0.002000 | Val Top-1: 58.14% | Val Top-5: 81.85%


Epoch [103/200] Loss: 1.6115 | LR: 0.002000 | Val Top-1: 57.88% | Val Top-5: 81.82%


Epoch [104/200] Loss: 1.5938 | LR: 0.002000 | Val Top-1: 57.26% | Val Top-5: 82.07%


Epoch [105/200] Loss: 1.5984 | LR: 0.002000 | Val Top-1: 58.01% | Val Top-5: 82.06%


Epoch [106/200] Loss: 1.5923 | LR: 0.002000 | Val Top-1: 57.48% | Val Top-5: 81.83%


Epoch [107/200] Loss: 1.5812 | LR: 0.002000 | Val Top-1: 57.56% | Val Top-5: 81.93%


Epoch [108/200] Loss: 1.5879 | LR: 0.002000 | Val Top-1: 57.38% | Val Top-5: 81.76%


Epoch [109/200] Loss: 1.5738 | LR: 0.002000 | Val Top-1: 57.79% | Val Top-5: 81.92%


Epoch [110/200] Loss: 1.5724 | LR: 0.002000 | Val Top-1: 57.76% | Val Top-5: 81.96%


Epoch [111/200] Loss: 1.5655 | LR: 0.002000 | Val Top-1: 57.72% | Val Top-5: 81.72%


Epoch [112/200] Loss: 1.5772 | LR: 0.002000 | Val Top-1: 57.58% | Val Top-5: 81.56%


Epoch [113/200] Loss: 1.5646 | LR: 0.002000 | Val Top-1: 58.16% | Val Top-5: 81.96%


Epoch [114/200] Loss: 1.5607 | LR: 0.002000 | Val Top-1: 58.16% | Val Top-5: 81.78%


Epoch [115/200] Loss: 1.5477 | LR: 0.002000 | Val Top-1: 57.74% | Val Top-5: 81.76%


Epoch [116/200] Loss: 1.5579 | LR: 0.002000 | Val Top-1: 57.93% | Val Top-5: 81.78%


Epoch [117/200] Loss: 1.5560 | LR: 0.002000 | Val Top-1: 57.82% | Val Top-5: 82.02%


Epoch [118/200] Loss: 1.5523 | LR: 0.002000 | Val Top-1: 57.90% | Val Top-5: 81.95%


Epoch [119/200] Loss: 1.5480 | LR: 0.002000 | Val Top-1: 58.27% | Val Top-5: 82.09%


Epoch [120/200] Loss: 1.5426 | LR: 0.000400 | Val Top-1: 58.53% | Val Top-5: 81.82%


Epoch [121/200] Loss: 1.4983 | LR: 0.000400 | Val Top-1: 58.86% | Val Top-5: 82.32%


Epoch [122/200] Loss: 1.4848 | LR: 0.000400 | Val Top-1: 58.96% | Val Top-5: 82.47%


Epoch [123/200] Loss: 1.4722 | LR: 0.000400 | Val Top-1: 58.86% | Val Top-5: 82.58%


Epoch [124/200] Loss: 1.4705 | LR: 0.000400 | Val Top-1: 58.92% | Val Top-5: 82.52%


Epoch [125/200] Loss: 1.4604 | LR: 0.000400 | Val Top-1: 59.12% | Val Top-5: 82.55%


Epoch [126/200] Loss: 1.4603 | LR: 0.000400 | Val Top-1: 59.36% | Val Top-5: 82.40%


Epoch [127/200] Loss: 1.4552 | LR: 0.000400 | Val Top-1: 59.29% | Val Top-5: 82.47%


Epoch [128/200] Loss: 1.4549 | LR: 0.000400 | Val Top-1: 59.35% | Val Top-5: 82.59%


Epoch [129/200] Loss: 1.4503 | LR: 0.000400 | Val Top-1: 59.02% | Val Top-5: 82.48%


Epoch [130/200] Loss: 1.4443 | LR: 0.000400 | Val Top-1: 59.07% | Val Top-5: 82.58%


Epoch [131/200] Loss: 1.4486 | LR: 0.000400 | Val Top-1: 59.21% | Val Top-5: 82.59%


Epoch [132/200] Loss: 1.4432 | LR: 0.000400 | Val Top-1: 59.12% | Val Top-5: 82.48%


Epoch [133/200] Loss: 1.4407 | LR: 0.000400 | Val Top-1: 59.28% | Val Top-5: 82.68%


Epoch [134/200] Loss: 1.4440 | LR: 0.000400 | Val Top-1: 59.42% | Val Top-5: 82.31%


Epoch [135/200] Loss: 1.4352 | LR: 0.000400 | Val Top-1: 59.25% | Val Top-5: 82.82%


Epoch [136/200] Loss: 1.4449 | LR: 0.000400 | Val Top-1: 59.15% | Val Top-5: 82.52%


Epoch [137/200] Loss: 1.4327 | LR: 0.000400 | Val Top-1: 59.07% | Val Top-5: 82.63%


Epoch [138/200] Loss: 1.4308 | LR: 0.000400 | Val Top-1: 59.38% | Val Top-5: 82.50%


Epoch [139/200] Loss: 1.4321 | LR: 0.000400 | Val Top-1: 59.34% | Val Top-5: 82.57%


Epoch [140/200] Loss: 1.4277 | LR: 0.000400 | Val Top-1: 59.23% | Val Top-5: 82.55%


Epoch [141/200] Loss: 1.4281 | LR: 0.000400 | Val Top-1: 59.26% | Val Top-5: 82.60%


Epoch [142/200] Loss: 1.4265 | LR: 0.000400 | Val Top-1: 59.33% | Val Top-5: 82.51%


Epoch [143/200] Loss: 1.4177 | LR: 0.000400 | Val Top-1: 59.30% | Val Top-5: 82.48%


Epoch [144/200] Loss: 1.4257 | LR: 0.000400 | Val Top-1: 59.21% | Val Top-5: 82.45%


Epoch [145/200] Loss: 1.4206 | LR: 0.000400 | Val Top-1: 59.34% | Val Top-5: 82.63%


Epoch [146/200] Loss: 1.4210 | LR: 0.000400 | Val Top-1: 59.31% | Val Top-5: 82.54%


Epoch [147/200] Loss: 1.4174 | LR: 0.000400 | Val Top-1: 59.54% | Val Top-5: 82.55%


Epoch [148/200] Loss: 1.4183 | LR: 0.000400 | Val Top-1: 59.33% | Val Top-5: 82.50%


Epoch [149/200] Loss: 1.4057 | LR: 0.000400 | Val Top-1: 59.34% | Val Top-5: 82.37%


Epoch [150/200] Loss: 1.4187 | LR: 0.000400 | Val Top-1: 59.37% | Val Top-5: 82.58%


Epoch [151/200] Loss: 1.4104 | LR: 0.000400 | Val Top-1: 59.33% | Val Top-5: 82.45%


Epoch [152/200] Loss: 1.4146 | LR: 0.000400 | Val Top-1: 59.02% | Val Top-5: 82.39%


Epoch [153/200] Loss: 1.4054 | LR: 0.000400 | Val Top-1: 59.51% | Val Top-5: 82.42%


Epoch [154/200] Loss: 1.4066 | LR: 0.000400 | Val Top-1: 59.46% | Val Top-5: 82.28%


Epoch [155/200] Loss: 1.4065 | LR: 0.000400 | Val Top-1: 59.56% | Val Top-5: 82.35%


Epoch [156/200] Loss: 1.4009 | LR: 0.000400 | Val Top-1: 59.62% | Val Top-5: 82.43%


Epoch [157/200] Loss: 1.4068 | LR: 0.000400 | Val Top-1: 59.45% | Val Top-5: 82.44%


Epoch [158/200] Loss: 1.4014 | LR: 0.000400 | Val Top-1: 59.54% | Val Top-5: 82.38%


Epoch [159/200] Loss: 1.3975 | LR: 0.000400 | Val Top-1: 59.25% | Val Top-5: 82.41%


Epoch [160/200] Loss: 1.4002 | LR: 0.000080 | Val Top-1: 59.53% | Val Top-5: 82.58%


Epoch [161/200] Loss: 1.3866 | LR: 0.000080 | Val Top-1: 59.28% | Val Top-5: 82.56%


Epoch [162/200] Loss: 1.3907 | LR: 0.000080 | Val Top-1: 59.54% | Val Top-5: 82.61%


Epoch [163/200] Loss: 1.3928 | LR: 0.000080 | Val Top-1: 59.51% | Val Top-5: 82.57%


Epoch [164/200] Loss: 1.3776 | LR: 0.000080 | Val Top-1: 59.60% | Val Top-5: 82.49%


Epoch [165/200] Loss: 1.3896 | LR: 0.000080 | Val Top-1: 59.66% | Val Top-5: 82.51%


Epoch [166/200] Loss: 1.3897 | LR: 0.000080 | Val Top-1: 59.62% | Val Top-5: 82.28%


Epoch [167/200] Loss: 1.3832 | LR: 0.000080 | Val Top-1: 59.50% | Val Top-5: 82.44%


Epoch [168/200] Loss: 1.3842 | LR: 0.000080 | Val Top-1: 59.41% | Val Top-5: 82.60%


Epoch [169/200] Loss: 1.3766 | LR: 0.000080 | Val Top-1: 59.56% | Val Top-5: 82.55%


Epoch [170/200] Loss: 1.3904 | LR: 0.000080 | Val Top-1: 59.53% | Val Top-5: 82.65%


Epoch [171/200] Loss: 1.3816 | LR: 0.000080 | Val Top-1: 59.58% | Val Top-5: 82.50%


Epoch [172/200] Loss: 1.3834 | LR: 0.000080 | Val Top-1: 59.68% | Val Top-5: 82.77%


Epoch [173/200] Loss: 1.3781 | LR: 0.000080 | Val Top-1: 59.48% | Val Top-5: 82.44%


Epoch [174/200] Loss: 1.3755 | LR: 0.000080 | Val Top-1: 59.65% | Val Top-5: 82.65%


Epoch [175/200] Loss: 1.3715 | LR: 0.000080 | Val Top-1: 59.62% | Val Top-5: 82.50%


Epoch [176/200] Loss: 1.3707 | LR: 0.000080 | Val Top-1: 59.83% | Val Top-5: 82.60%


Epoch [177/200] Loss: 1.3703 | LR: 0.000080 | Val Top-1: 59.34% | Val Top-5: 82.59%


Epoch [178/200] Loss: 1.3795 | LR: 0.000080 | Val Top-1: 59.30% | Val Top-5: 82.55%


Epoch [179/200] Loss: 1.3731 | LR: 0.000080 | Val Top-1: 59.47% | Val Top-5: 82.46%


Epoch [180/200] Loss: 1.3754 | LR: 0.000080 | Val Top-1: 59.65% | Val Top-5: 82.41%


Epoch [181/200] Loss: 1.3761 | LR: 0.000080 | Val Top-1: 59.61% | Val Top-5: 82.62%


Epoch [182/200] Loss: 1.3749 | LR: 0.000080 | Val Top-1: 59.28% | Val Top-5: 82.70%


Epoch [183/200] Loss: 1.3789 | LR: 0.000080 | Val Top-1: 59.57% | Val Top-5: 82.65%


Epoch [184/200] Loss: 1.3801 | LR: 0.000080 | Val Top-1: 59.44% | Val Top-5: 82.51%


Epoch [185/200] Loss: 1.3803 | LR: 0.000080 | Val Top-1: 59.57% | Val Top-5: 82.64%


Epoch [186/200] Loss: 1.3686 | LR: 0.000080 | Val Top-1: 59.74% | Val Top-5: 82.47%


Epoch [187/200] Loss: 1.3769 | LR: 0.000080 | Val Top-1: 59.38% | Val Top-5: 82.49%


Epoch [188/200] Loss: 1.3802 | LR: 0.000080 | Val Top-1: 59.65% | Val Top-5: 82.51%


Epoch [189/200] Loss: 1.3753 | LR: 0.000080 | Val Top-1: 59.56% | Val Top-5: 82.78%


Epoch [190/200] Loss: 1.3699 | LR: 0.000080 | Val Top-1: 59.57% | Val Top-5: 82.47%


Epoch [191/200] Loss: 1.3693 | LR: 0.000080 | Val Top-1: 59.45% | Val Top-5: 82.57%

Early stopping: No improvement for 15 epochs
Best accuracy: 59.83%
Fine-tuning completed. Best accuracy: 59.83%

EVALUATING FINAL MODEL




✓ Results saved to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/results/cifar100_grasp_results.json

TASK 1B PIPELINE COMPLETE
Final Top-1 Accuracy: 59.45%
Final Top-5 Accuracy: 82.57%
Final Sparsity: 95.20%
Model Size: 58.43 MB


CIFAR-100 GraSP Pruning Complete!
Best Accuracy: 59.83%


### Step 3: COO Sparse Conversion and Profiling

In [18]:
# Convert to COO sparse format (inherited from VGG16_Pruning)
cifar100_sparse_tensors, cifar100_memory_stats = cifar100_model.convert_to_coo_sparse(cifar100_masks)

# Verify consistency
cifar100_model.verify_mask_coo_consistency(cifar100_masks, cifar100_sparse_tensors)


CONVERTING TO COO SPARSE FORMAT

features.0:
  Total params: 1,728
  Non-zeros: 193
  Sparsity: 88.83%
  Dense size: 0.01 MB
  Sparse size: 0.01 MB
  Compression: 0.99x

features.3:
  Total params: 36,864
  Non-zeros: 4,005
  Sparsity: 89.14%
  Dense size: 0.14 MB
  Sparse size: 0.14 MB
  Compression: 1.02x

features.7:
  Total params: 73,728
  Non-zeros: 8,525
  Sparsity: 88.44%
  Dense size: 0.28 MB
  Sparse size: 0.29 MB
  Compression: 0.96x

features.10:
  Total params: 147,456
  Non-zeros: 14,630
  Sparsity: 90.08%
  Dense size: 0.56 MB
  Sparse size: 0.50 MB
  Compression: 1.12x

features.14:
  Total params: 294,912
  Non-zeros: 31,077
  Sparsity: 89.46%
  Dense size: 1.12 MB
  Sparse size: 1.07 MB
  Compression: 1.05x

features.17:
  Total params: 589,824
  Non-zeros: 61,346
  Sparsity: 89.60%
  Dense size: 2.25 MB
  Sparse size: 2.11 MB
  Compression: 1.07x

features.20:
  Total params: 589,824
  Non-zeros: 49,204
  Sparsity: 91.66%
  Dense size: 2.25 MB
  Sparse size: 1.69 MB

{'features.0': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 1728},
 'features.3': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 36864},
 'features.7': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 73728},
 'features.10': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 147456},
 'features.14': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 294912},
 'features.17': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 589824},
 'features.20': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 589824},
 'features.24': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 1179648},
 'features.27': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 2359296},
 'features.30': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 2359296},
 'features.34': {'consistent': True,
  'num_mismatches': 0,
  'total_elements': 2359296},
 'features.37': {'consis

In [19]:
# Profile sparse model
cifar100_model.profile_sparse_model(
    sparse_tensors=cifar100_sparse_tensors,
    test_loader=test_loader_100,
    train_loader=train_loader_100,
    device=DEVICE
)


PROFILING SPARSE MODEL


REPLACING DENSE WEIGHTS WITH SPARSE TENSORS

✓ features.0: Replaced with sparse tensor (COO format)
✓ features.3: Replaced with sparse tensor (COO format)
✓ features.7: Replaced with sparse tensor (COO format)
✓ features.10: Replaced with sparse tensor (COO format)
✓ features.14: Replaced with sparse tensor (COO format)
✓ features.17: Replaced with sparse tensor (COO format)
✓ features.20: Replaced with sparse tensor (COO format)
✓ features.24: Replaced with sparse tensor (COO format)
✓ features.27: Replaced with sparse tensor (COO format)
✓ features.30: Replaced with sparse tensor (COO format)
✓ features.34: Replaced with sparse tensor (COO format)
✓ features.37: Replaced with sparse tensor (COO format)
✓ features.40: Replaced with sparse tensor (COO format)
✓ classifier.0: Replaced with sparse tensor (COO format)
✓ classifier.3: Replaced with sparse tensor (COO format)
✓ classifier.6: Replaced with sparse tensor (COO format)

✓ All weights replaced with spar

### Step 4: Save Final Model

In [20]:
# Save final pruned model
cifar100_final_path = os.path.join(MODELS_DIR, 'task1b_cifar100_final.pt')
cifar100_model.save(cifar100_final_path)

print(f"CIFAR-100 final model saved to: {cifar100_final_path}")

Saving model to /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/models/task1b_cifar100_final.pt
Model saved successfully
CIFAR-100 final model saved to: /content/drive/MyDrive/AI624-Edge-Devices/pa1.1/task1b/models/task1b_cifar100_final.pt


## Task 1(b): GraSP Saliency-Based Iterative Pruning


#### CIFAR-10 GraSP Results

| Metric                          | Baseline  | After GraSP (80% sparsity) | Change from Baseline |
| ------------------------------- | --------- | -------------------------- | -------------------- |
| Model Size (MB) - Dense         | 58.25     | 58.14                      | -0.11 MB             |
| **Model Size (MB) - Sparse**    | **58.25** | **25.03**                  | **-33.22 MB**        |
| **Sparsity Ratio (%)**          | **0.00**  | **80.00**                  | **+80.00%**          |
| Test Top-1 Accuracy (%)         | 94.16     | 71.33                      | -22.83%              |
| Test Top-5 Accuracy (%)         | 99.71     | 97.65                      | -2.06%               |
| Train Top-1 Accuracy (%)        | 100.00    | 75.91                      | -24.09%              |
| Train Top-5 Accuracy (%)        | 100.00    | 99.58                      | -0.42%               |
| Latency (ms/batch)              | 12.24     | 13.58                      | +1.34 ms             |
| COO Compression Ratio           | 1.00x     | 2.32x                      | Dense → Sparse       |

#### CIFAR-100 GraSP Results

| Metric                          | Baseline  | After GraSP (80% sparsity) | Change from Baseline |
| ------------------------------- | --------- | -------------------------- | -------------------- |
| Model Size (MB) - Dense         | 58.43     | 58.31                      | -0.12 MB             |
| **Model Size (MB) - Sparse**    | **58.43** | **24.71**                  | **-33.72 MB**        |
| **Sparsity Ratio (%)**          | **0.00**  | **80.00**                  | **+80.00%**          |
| Test Top-1 Accuracy (%)         | 74.00     | 59.45                      | -14.55%              |
| Test Top-5 Accuracy (%)         | 90.56     | 82.57                      | -7.99%               |
| Train Top-1 Accuracy (%)        | 99.93     | 76.78                      | -23.15%              |
| Train Top-5 Accuracy (%)        | 100.00    | 95.26                      | -4.74%               |
| Latency (ms/batch)              | 6.60      | 3.93                       | -2.67 ms             |
| COO Compression Ratio           | 1.00x     | 2.36x                      | Dense → Sparse       |

### Analysis: Expected vs. Observed Changes

#### 1. Model Size

**Expected:**
- Dense format: unchanged (zeroed weights still occupy memory)
- COO sparse format: ~2.3x compression at 80% sparsity

**Observed:**
- Dense: 58.14 MB → 58.14 MB (no change, as expected)
- Sparse: 58.14 MB → 25.03 MB (2.32x compression for CIFAR-10)
- Sparse: 58.31 MB → 24.71 MB (2.36x compression for CIFAR-100)

#### 2. Accuracy Compared to Baseline

**Expected:**
- GraSP training from random initialization would achieve lower accuracy than pretrained model
- GraSP's gradient-preserving criterion should minimize accuracy loss

**Observed:**
- CIFAR-10: 94.16% → 71.33% = **22.83% drop** (exceeds 15% target)
- CIFAR-100: 74.00% → 59.45% = **14.55% drop** (within target)

**Explanation:**
- Training from random initialization (not pretrained) significantly impacts final accuracy
- GraSP's iterative approach (40%→60%→80%) allows gradual adaptation
- CIFAR-10's higher baseline makes absolute accuracy drop more visible

### Discussion:

#### Sparse Kernel Usage

**Question:** Are sparse inference optimization kernels being used under the hood?

**Answer:** **No, PyTorch does not use sparse kernels for our unstructured pruning.**

### SNIP vs. GraSP Comparison

#### Algorithm Comparison

**SNIP (Single-shot Network Pruning):**
- Saliency: Connection sensitivity `s = |θ · ∂L/∂θ|`
- Single pruning event before training
- Fast but less adaptive

**GraSP (Gradient Signal Preservation):**
- Saliency: Hessian-gradient product `Hg = ∇(g^T · g)`
- Iterative pruning during training
- Captures weight interactions via second-order gradients

#### Would SNIP Give Better Results?

**Expected:** No, GraSP should perform better for this task.

**Why:**
1. **Training from scratch:** GraSP's iterative approach (40%→60%→80%) allows the network to adapt gradually
2. **Second-order information:** GraSP's Hessian-gradient product captures how weights interact, not just individual importance
3. **Gradient flow preservation:** GraSP explicitly preserves gradient paths, critical for training from random initialization
4. **SNIP limitation:** Single-shot pruning before training doesn't account for how weights evolve during learning

**Verdict:** GraSP's iterative, gradient-aware approach is better suited for training from random initialization with high sparsity. SNIP would likely achieve lower final accuracy.